{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "ppBlnNxt5esa",
        "lfs6AHGhaTsr"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "The goal of this section is two-fold.\n",
        "\n",
        "First, seeing which genomes contain the flanking regions/genes of our genes of interest——meaning, in which genomes are the flanking genes present at a reasonable distance.\n",
        "\n",
        "Second, trying to detect non-coding homologs for our genes of interest by running a sensitive search of the region bounded by the flanking genes."
      ],
      "metadata": {
        "id": "rYdonkDuP7W_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Identifying flanking genes and regions**"
      ],
      "metadata": {
        "id": "f6SYs-uAXl9K"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Our genes of interest aren't strictly genes, but representative members of gene families. Other members of their family might have different flanking regions.\n",
        "\n",
        "To account for this possibility, we define \"flanking genes\" as the unique set of genes that surround the members of our gene family of interest.\n",
        "\n",
        "To this end, we proceed in the following way:\n",
        "\n",
        "1. First, identify all flanking genes, i.e. immediate upstream and downstream genes, for each member of the gene family in question\n",
        "2. Replace the upstream and downstream gene IDs with their corresponding family/cluster ID\n",
        "3. Extract the set of unique pairs of gene family IDs from this list\n",
        "\n",
        "In addition to extracting flanking genes, we also extract sequences immediately upstream and downstream of our genes of interest. To this end, we choose (randomly) one of the genomes in which both flanks are present, and extract +/-500bp from gene edge. This is repeated for each pair of gene family IDs."
      ],
      "metadata": {
        "id": "cZuCcatqXq-6"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **1. Setup**"
      ],
      "metadata": {
        "id": "MiB-_G6PjH7Y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#For the next step, the mmseqs2 cluster names in the protein-cluster designation file needs to be replaced with the mmseqs-silix one\n",
        "\n",
        "awk 'FNR==NR {rep[$2]=$1; next} {$1 = ($1 in rep ? rep[$1] : $1); print}' mmseqs_silix_reclustering/replacements.tsv Ecoli.all.filtered.prot.clusters.longestrepresentative.tsv > Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.tsv\n",
        "cut -f1,2 -d \" \" ../Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.tsv | tail -n+2 | awk '{print $2\"\\t\"$1}' > Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.formatted.tsv\n",
        "\n",
        "#Furthermore, all proteins——including those three categories excluded from analyses at an earlier step——need to be clustered as well\n",
        "#In my previous code I did the clustering with usearch, which is another similarity-based clustering tool\n",
        "#This is fine for our purposes\n",
        "\n",
        "usearch -sortbylength Ecoli.all.prot.pangenomedb.faa -fastaout Ecoli.all.prot.pangenomedb.sorted.faa -minseqlength 1\n",
        "usearch -cluster_smallmem Ecoli.all.prot.pangenomedb.sorted.faa -id 0.8 -centroids Ecoli.all.prot.nr.faa -uc Ecoli.all.clusters.uc\n",
        "\n",
        "#Convert the .uc to a a clustered tsv\n",
        "cut -f9,10 Ecoli.all.clusters.uc | sort -u | awk -F '\\t' '($2==\"*\")' | awk -F '\\t' '{print $1\"\\t\"$1}' > clusters.tsv\n",
        "cut -f9,10 Ecoli.all.clusters.uc | sort -u | awk -F '\\t' '($2!=\"*\")' | awk -F '\\t' '{print $1\"\\t\"$2}' >> clusters.tsv\n",
        "\n",
        "#The clusters.tsv file is a two-column gene-cluster file, similar to the one generated by mmseqs2\n"
      ],
      "metadata": {
        "id": "B5ugkVRY28MB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Debugging"
      ],
      "metadata": {
        "id": "ppBlnNxt5esa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#I don't know why I needed to make an all-protein database of the sort seen below\n",
        "#This could either be required, or be a vestige of old code I had.\n",
        "#I can't think of a reason why I would need to generate this alt protein file, but in case it's important: the code is retained below\n",
        "\n",
        "cd ../Ecoli_gtfs\n",
        "for i in $(ls *gtf | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "cat \"$i\".nr.gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\"$i\".fasta -bed - > \"$i\".CDS.faa\n",
        "done\n",
        "\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_cds\n",
        "mv *CDS.faa /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_cds/\n",
        "\n",
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_cds/\n",
        "for i in $(ls *faa | rev | cut -f3- -d '.' | rev)\n",
        "do\n",
        "/stor/work/Ochman/hassan/tools/faTrans -stop \"$i\".CDS.faa \"$i\".prot.faa\n",
        "done\n",
        "\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_prot\n",
        "mv *prot.faa /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_prot\n",
        "\n",
        "#prep done\n",
        "\n",
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/\n",
        "\n",
        "#Just take the 450 Ecoli genomes with good lineages\n",
        "for i in $(awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | cut -f1)\n",
        "do\n",
        "sed \"s/>/>\"$i\"@/g\" Ecoli_prot/\"$i\".prot.faa >> Ecoli.all.prot.faa\n",
        "sed \"s/>/>\"$i\"@/g\" Ecoli_cds/\"$i\".CDS.faa >> Ecoli.all.CDS.faa\n",
        "done\n",
        "\n",
        "grep \"^>\" Ecoli.all.prot.nr.faa | grep --no-group-separator -F -A1 -f - Ecoli.all.CDS.faa > Ecoli.all.CDS.nr.faa\n",
        "\n",
        "less Ecoli.all.prot.pangenomedb.faa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 108
        },
        "collapsed": true,
        "id": "xWzEodxP5d3Q",
        "outputId": "0aa44169-cf0e-42b5-ff77-f4e0d6a7dcc2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "invalid decimal literal (ipython-input-3165409109.py, line 24)",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipython-input-3165409109.py\"\u001b[0;36m, line \u001b[0;32m24\u001b[0m\n\u001b[0;31m    for i in $(awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | cut -f1)\u001b[0m\n\u001b[0m                                                                                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid decimal literal\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **2. Making gene flanks tables for each gene of interest**"
      ],
      "metadata": {
        "id": "b6VTSTbl5iAz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#For subsequent steps, the gtfs need to be sorted\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/*nr.gtf)\n",
        "do\n",
        "bedtools sort -i $i > temp && mv temp $i\n",
        "done\n",
        "\n",
        "mkdir flank\n",
        "cp Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.formatted.tsv flank\n",
        "cd flank\n",
        "\n",
        "#The next bit of script progresses in a few steps\n",
        "#1. For each member in a gene family of interest, get its upstream and downstream gene\n",
        "#2. Make a five-column tab-separated that lists the gene of interest, upstream gene, its cluster, downstream gene, its cluster\n",
        "#3. Concatenate these rows, each corresponding to a member of the gene familiy, into a single file\n",
        "#4. If any row contains fewer than five columns, that means that gene does not have both upstream and downstream flanking genes\n",
        "#5. Retain each unique pair of family name for the upstream and downstream genes\n",
        "\n",
        "#For each gene in the list of genes of interest (e.g., ORFan candidates)\n",
        "#We loop over each of the members in their family\n",
        "#Since the code is time-consuming, I put it in an \"echo\" wrapper below\n",
        "#I'm showing an example for one gene of interest to demonstrate how the code looks\n",
        "for i in $(grep \"JSKL00000000@EMDEOKIC_02191(\" Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.formatted.tsv | cut -f1 -d \"(\")\n",
        "do\n",
        "    genome_name=$(echo $i | cut -f1 -d \"@\"); ortholog_name=$(echo $i | cut -f2 -d \"@\"); #extract genome name from each gene ID\n",
        "    gene_ortholog_gtf=$(grep -w $ortholog_name ../../Ecoli_gtfs/$genome_name.nr.gtf) #extract gtf line corresponding to that gene\n",
        "    echo $i > JSKL00000000@EMDEOKIC_02191_temp #Start building five-column row. First column——gene name\n",
        "    #Get the upstream gene\n",
        "    echo $gene_ortholog_gtf | sed \"s/ /%/9\" | sed \"s/ /%/9\" | sed \"s/ /\\t/g\" | sed \"s/%/ /g\" | bedtools closest -a - -b ../../Ecoli_gtfs/$genome_name.nr.gtf -D a -id -io | awk -F '\\t' '($12==\"CDS\")' | cut -f 6 -d \"\\\"\" | sed \"s/^/$genome_name@/g\" >> \"$i\"_upstream\n",
        "    #Add it to the growing table\n",
        "    cat \"$i\"_upstream >> JSKL00000000@EMDEOKIC_02191_temp\n",
        "    #Add its corresponding gene family/cluster designation as well\n",
        "    grep -F -f \"$i\"_upstream ../../clusters.tsv | cut -f2 >> JSKL00000000@EMDEOKIC_02191_temp\n",
        "    rm \"$i\"_upstream\n",
        "    #Get the downstream gene\n",
        "    echo $gene_ortholog_gtf | sed \"s/ /%/9\" | sed \"s/ /%/9\" | sed \"s/ /\\t/g\" | sed \"s/%/ /g\" | bedtools closest -a - -b ../../Ecoli_gtfs/$genome_name.nr.gtf -D a -iu -io | awk -F '\\t' '($12==\"CDS\")' | cut -f 6 -d \"\\\"\" | sed \"s/^/$genome_name@/g\" >> \"$i\"_downstream\n",
        "    #Add it to the growing table\n",
        "    cat \"$i\"_downstream >> JSKL00000000@EMDEOKIC_02191_temp\n",
        "    #Add its corresponding gene family/cluster designation as well\n",
        "    grep -F -f \"$i\"_downstream ../../clusters.tsv | cut -f2 >> JSKL00000000@EMDEOKIC_02191_temp\n",
        "    rm \"$i\"_downstream\n",
        "    if [ \"$(wc -l < JSKL00000000@EMDEOKIC_02191_temp)\" -eq 5 ]; then #Only consider cases in which the file has 5 rows\n",
        "        paste -sd, JSKL00000000@EMDEOKIC_02191_temp | sed \"s/,/\\t/g\" >> JSKL00000000@EMDEOKIC_02191_flank_analysis #Transform into five-column table\n",
        "        grep -E '^([^@]*@){5}[^@]*$' JSKL00000000@EMDEOKIC_02191_flank_analysis | cut -f3,5 | sort -u > JSKL00000000@EMDEOKIC_02191_flanks #Retain the family ID of each unique pair of flanking genes\n",
        "    fi\n",
        "done\n",
        "\n",
        "#Parallelized code:\n",
        "\n",
        "for j in $(cut -f1 -d \"(\" ../step1_genusspecific_ORFan.replaced.txt)\n",
        "do\n",
        "    # Create the script file for each $j\n",
        "    echo \"#!/bin/bash\" > \"${j}.sh\"  # Start each script with a shebang\n",
        "\n",
        "    # Generate each line of the original script within an echo statement and append to the script file\n",
        "    echo \"for i in \\$(grep \\\"$j(\\\" Ecoli.all.filtered.prot.clusters.longestrepresentative.replaced.formatted.tsv | cut -f1 -d \\\"(\\\")\" >> \"${j}.sh\"  # Loop over clusters\n",
        "    echo \"do\" >> \"${j}.sh\"\n",
        "    echo \"    genome_name=\\$(echo \\$i | cut -f1 -d \\\"@\\\"); ortholog_name=\\$(echo \\$i | cut -f2 -d \\\"@\\\");\" >> \"${j}.sh\"\n",
        "    echo \"    gene_ortholog_gtf=\\$(grep -w \\$ortholog_name ../../Ecoli_gtfs/\\$genome_name.nr.gtf)\" >> \"${j}.sh\"\n",
        "    echo \"    echo \\$i > ${j}_temp\" >> \"${j}.sh\"\n",
        "    echo \"    echo \\$gene_ortholog_gtf | sed \\\"s/ /%/9\\\" | sed \\\"s/ /%/9\\\" | sed \\\"s/ /\\\\t/g\\\" | sed \\\"s/%/ /g\\\" | bedtools closest -a - -b ../../Ecoli_gtfs/\\$genome_name.nr.gtf -D a -id -io | awk -F '\\\\t' '(\\$12==\\\"CDS\\\")' | cut -f 6 -d \\\"\\\\\\\"\\\" | sed \\\"s/^/\\$genome_name@/g\\\" >> \\\"\\$i\\\"_upstream\" >> \"${j}.sh\"\n",
        "    echo \"    cat \\\"\\$i\\\"_upstream >> ${j}_temp\" >> \"${j}.sh\"\n",
        "    echo \"    grep -F -f \\\"\\$i\\\"_upstream ../../clusters.tsv | cut -f2 >> ${j}_temp\" >> \"${j}.sh\"\n",
        "    echo \"    rm \\\"\\$i\\\"_upstream\" >> \"${j}.sh\"\n",
        "    echo \"    echo \\$gene_ortholog_gtf | sed \\\"s/ /%/9\\\" | sed \\\"s/ /%/9\\\" | sed \\\"s/ /\\\\t/g\\\" | sed \\\"s/%/ /g\\\" | bedtools closest -a - -b ../../Ecoli_gtfs/\\$genome_name.nr.gtf -D a -iu -io | awk -F '\\\\t' '(\\$12==\\\"CDS\\\")' | cut -f 6 -d \\\"\\\\\\\"\\\" | sed \\\"s/^/\\$genome_name@/g\\\" >> \\\"\\$i\\\"_downstream\" >> \"${j}.sh\"\n",
        "    echo \"    cat \\\"\\$i\\\"_downstream >> ${j}_temp\" >> \"${j}.sh\"\n",
        "    echo \"    grep -F -f \\\"\\$i\\\"_downstream ../../clusters.tsv | cut -f2 >> ${j}_temp\" >> \"${j}.sh\"\n",
        "    echo \"    rm \\\"\\$i\\\"_downstream\" >> \"${j}.sh\"\n",
        "\n",
        "    # Add the conditional part for checking the number of lines\n",
        "    echo \"    if [ \\\"\\$(wc -l < ${j}_temp)\\\" -eq 5 ]; then\" >> \"${j}.sh\"\n",
        "    echo \"        paste -sd, ${j}_temp | sed \\\"s/,/\\\\t/g\\\" >> ${j}_flank_analysis\" >> \"${j}.sh\"\n",
        "    echo \"        grep -E '^([^@]*@){5}[^@]*\\$' ${j}_flank_analysis | cut -f3,5 | sort -u > ${j}_flanks\" >> \"${j}.sh\"\n",
        "    echo \"    fi\" >> \"${j}.sh\"\n",
        "\n",
        "    # Close the loop\n",
        "    echo \"done\" >> \"${j}.sh\"\n",
        "done"
      ],
      "metadata": {
        "id": "dBvAKdoEjWf6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **3. Extract flanking genes (geneflanks) and flanking regions (proxflanks)**"
      ],
      "metadata": {
        "id": "NuOXOwPJjbY7"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 144
        },
        "collapsed": true,
        "id": "W2JwPm0Y4YSO",
        "outputId": "a86e90ba-bee8-472f-8ebe-6feddbd5207e"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "invalid decimal literal (ipython-input-3587605900.py, line 102)",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipython-input-3587605900.py\"\u001b[0;36m, line \u001b[0;32m102\u001b[0m\n\u001b[0;31m    blastn -query geneflanks.faa -db /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 109734 -max_hsps 1 -out geneflanks_extragenus_blastn\u001b[0m\n\u001b[0m                                                                       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid decimal literal\n"
          ]
        }
      ],
      "source": [
        "#Make a helper file concatenating the sequences of all flanking genes so next steps go quicker\n",
        "cat *_flanks | sed \"s/\\t/\\n/g\" | sort -u | grep -A1 --no-group-separator -F -f - ../../Ecoli.all.CDS.faa > all_flanks.CDS.faa\n",
        "\n",
        "#Geneflanks:\n",
        "\n",
        "#Extracting flanking genes is straightforward\n",
        "#For each unique flanking gene pair, assign a line number\n",
        "#Extract their CDSs\n",
        "#The identifier should be of the form: GeneOfInterest_linenumber_up/down_FlankingGeneFamily\n",
        "\n",
        "for j in $(ls *_flanks | rev | cut -f2- -d \"_\" | rev)\n",
        "do\n",
        "for i in $(cat -n \"$j\"_flanks | sed \"s/^ *//g\" | sed \"s/\\t/,/g\")\n",
        "do\n",
        "linenumber=$(echo $i | cut -f1 -d ',')\n",
        "echo $i | cut -f2 -d ',' | grep --no-group-separator -A1 -f - all_flanks.CDS.faa | sed \"s/>/>\"$j\"_\"$linenumber\"_up_/g\" >> geneflanks.faa\n",
        "echo $i | cut -f3 -d ',' | grep --no-group-separator -A1 -f - all_flanks.CDS.faa | sed \"s/>/>\"$j\"_\"$linenumber\"_down_/g\" >> geneflanks.faa\n",
        "done\n",
        "done\n",
        "\n",
        "#Proxflanks:\n",
        "\n",
        "#To get sequences immediately upstream and downstream of the gene of interest:\n",
        "#Identify a genome——any genome——just pick one at random——where the gene has both flanking genes\n",
        "#Extract +/-500bp\n",
        "\n",
        "for j in $(ls *_flanks | rev | cut -f2- -d \"_\" | rev)\n",
        "do\n",
        "#The next line takes one ortholog at random from each unique pair of flanking genes\n",
        "for i in $(grep -E '^([^@]*@){5}[^@]*$' \"$j\"_flank_analysis | awk -F '\\t' '{print $0,$3\"_\"$5}' | awk '!seen[$6]++' | cat -n | sed \"s/^ *//g\" | sed \"s/\\t/,/g\" | cut -f1 -d \" \")\n",
        "do\n",
        "linenumber=$(echo $i | cut -f1 -d ',')\n",
        "ortholog_name=$(echo $i | cut -f2 -d ',' | cut -f2 -d \"@\" | cut -f1 -d \"(\")\n",
        "genome_name=$(echo $i | cut -f2 -d ',' | cut -f1 -d \"@\")\n",
        "grep -w $ortholog_name ../../Ecoli_gtfs/$genome_name.filtered.nr.gtf | sed \"s/ \\\"/ \\\"\"$j\"_\"$linenumber\"_/g\" >> proxflanks_interim.gtf #Get a concatenated gtf\n",
        "done\n",
        "done\n",
        "\n",
        "#Get the sequences\n",
        "#The next line is required to extract the gene sequences from the gtf made above\n",
        "#Since the gtf has genes from a bunch of genomes (indicated by its harboring contig in column 1), the multifasta identifier has to have contigs\n",
        "seqkit fx2tab ../../Ecoli.all.genomes.faa | sed \"s/\\t$//g\" | cut -f2- -d \"@\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" > ../../Ecoli.all.genomes.temp.faa\n",
        "#The $4>0 is part is to get rid of those cases in which the start codon dips to negative\n",
        "#I'm making the choice to skip those that don't have 500bp on either edge\n",
        "cat proxflanks_interim.gtf | awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$4-500,$4,$6,$7,$8,$9}' |  awk -F '\\t' '($4>0)' | sed \"s/ \\\"/ \\\"left500_/g\" | gtf2bed | bedtools getfasta -s -name -fi ../../Ecoli.all.genomes.temp.faa -bed - > proxflanks.faa\n",
        "cat proxflanks_interim.gtf | awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$5,$5+500,$6,$7,$8,$9}' | sed \"s/ \\\"/ \\\"right500_/g\" | gtf2bed | bedtools getfasta -s -name -fi ../../Ecoli.all.genomes.temp.faa -bed - >> proxflanks.faa\n",
        "\n",
        "#Remove those that lack either or both flanks:\n",
        "#The code below counts the number of commas per line after massaging it, and if there are no commas——that means at least one flank is missing\n",
        "grep \"^>\" proxflanks.faa | sed \"s/_/\\t/\" | awk -F'\\t' '{ values[$2] = (values[$2] == \"\" ? $1 : values[$2] \", \" $1) } END { for (value in values) { print value \"\\t\" values[value] } }' | grep \",\" | cut -f 1 -d \"(\" | grep --no-group-separator -A1 -f - proxflanks.faa > interim\n",
        "mv interim proxflanks.faa"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Identifying genomes which contain both flanks at a reasonable distance**"
      ],
      "metadata": {
        "id": "qaZ0J9eKzDNd"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Once we have the flank sequences for each of our genes, we search them (blastn) against our genome databases to identify those genomes in which both of them are present within a reasonable distance (no closer than half the length of the gene sequence, and no further than 10000bp). After attaching taxonomy information to these files, we'll know which genomes have the flanking regions of these genes. (I only measure this distance between the best hit for each flank sequence per genome)\n",
        "\n",
        "This is how the code below is written, but I want to generate an additional set of files that sets the length at 5X the gene length, and gets rid of the minimal limit."
      ],
      "metadata": {
        "id": "Zqdof6w5zYvd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#First, the blast searches (Time consuming)\n",
        "blastn -query geneflanks.faa -db /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 109734 -max_hsps 1 -out geneflanks_extragenus_blastn\n",
        "#109734\n",
        "blastn -query geneflanks.faa -db /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Ecoli_excluded.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 195119 -max_hsps 1 -out geneflanks_intragenus_blastn\n",
        "#195119\n",
        "blastn -query geneflanks.faa -db ../../Ecoli.all.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 86906 -max_hsps 1 -out geneflanks_pangenome_blastn\n",
        "#86906\n",
        "blastn -query proxflanks.faa -db /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 109734 -max_hsps 1 -out proxflanks_extragenus_blastn\n",
        "blastn -query proxflanks.faa -db /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Ecoli_excluded.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 195119 -max_hsps 1 -out proxflanks_intragenus_blastn\n",
        "blastn -query proxflanks.faa -db ../../Ecoli.all.genomes -outfmt '6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps sstrand qstart qend sstart send qlen slen evalue bitscore' -num_threads 104 -max_target_seqs 86906 -max_hsps 1 -out proxflanks_pangenome_blastn\n",
        "#choice: max-hsps 1\n",
        "\n",
        "#Make a list of genomes in which both flanks are found\n",
        "#No pid, qidentity or evalue cutoffs applied\n",
        "cat proxflanks_extragenus_blastn proxflanks_intragenus_blastn proxflanks_pangenome_blastn | cut -f-2 | rev | awk -F'\\t' 'BEGIN{OFS=FS} {split($2,a,\"_\"); $2=a[1]\"_\"a[2]\"\\t\"a[3]; for(i=4;i<=length(a);i++) $2=$2\"_\"a[i]; print}' | rev | sed \"s/_/\\t/1\" | awk -F '\\t' '{OFS=FS}{print $1,$2\"%\"$4}' | awk -F'\\t' '{ values[$2] = (values[$2] == \"\" ? $1 : values[$2] \", \" $1) } END { for (value in values) { print value \"\\t\" values[value] } }' | egrep \"left.*right|right.*left\" | cut -f1 | awk -F'%' '{ values[$1] = (values[$1] == \"\" ? $2 : values[$1] \", \" $2) } END { for (value in values) { print value \"\\t\" values[value] } }' > proxflanks_targets.txt\n",
        "cat geneflanks_extragenus_blastn geneflanks_intragenus_blastn geneflanks_pangenome_blastn | sed \"s/_up_/_up%/g\" | sed \"s/_down_/_down%/g\" | sed \"s/\\t/%/\" | cut -f1 | cut -f1,3 -d \"%\" | sed \"s/%/\\t/g\" | sed \"s/_up/\\tup/g\" | sed \"s/_down/\\tdown/g\" | awk -F '\\t' '{print $2\"\\t\"$1\">\"$3}' | sort -u | awk -F'\\t' '{ values[$2] = (values[$2] == \"\" ? $1 : values[$2] \", \" $1) } END { for (value in values) { print value \"\\t\" values[value] } }' | egrep \"up.*down|down.*up\" | cut -f1 | rev | sed \"s/>/%/1\" | rev | awk -F'%' '{ values[$1] = (values[$1] == \"\" ? $2 : values[$1] \", \" $2) } END { for (value in values) { print value \"\\t\" values[value] } }' > geneflanks_targets.txt\n",
        "\n",
        "#We know make files for each gene that lists each target genome in the form of a five-column table\n",
        "#contig name, start position, end position, length, strand\n",
        "for i in $(cut -f 1 proxflanks_targets.txt)\n",
        "do\n",
        "  echo \"grep -P \\\"$i\\t\\\" proxflanks_targets.txt | sed \\\"s/,/\\\\n/g\\\" | sed \\\"s/\\t/\\\\n/g\\\" | sed \\\"s/^ *//g\\\" | tail -n+2 > ${i}_proxflanks_targetlist.txt\"\n",
        "  echo \"cat proxflanks_extragenus_blastn proxflanks_intragenus_blastn proxflanks_pangenome_blastn | grep -w -F -f ${i}_proxflanks_targetlist.txt - | grep \\\"${i}_\\\" | cut -f2- -d \\\"_\\\" | sed 's/_/\\t/3' | cut -f1,3- | awk -F '\\t' '{OFS=\\\"\\\"}{print \\$13,\\\"%\\\",\\$14,\\\"%\\\",\\$10,\\\"\\t\\\",\\$2}' | awk -F'\\t' '{ values[\\$2] = (values[\\$2] == \\\"\\\" ? \\$1 : values[\\$2] \\\", \\\" \\$1) } END { for (value in values) { print value \\\"\\t\\\" values[value] } }' | sed \\\"s/%plus, /%/g\\\" | sed \\\"s/%minus, /%/g\\\" | sed \\\"s/%plus/\\tplus/g\\\" | sed \\\"s/%minus/\\tminus/g\\\" | sed \\\"s/%/,/g\\\" | sed \\\"s/\\t/,/g\\\" | sed \\\"s/ //g\\\" | awk -F',' '{identifier = \\$1; values = \\$2 \\\",\\\" \\$3 \\\",\\\" \\$4 \\\",\\\" \\$5; split(values, array, \\\",\\\"); asort(array); middle1 = array[2]; middle2 = array[3]; difference = middle2 - middle1; if (difference >= 0) { print identifier, middle1, middle2, difference, \\$6; } else { print identifier, middle2, middle1, -difference, \\$6; } }' > ${i}_proxflanks_intervalinfo\"\n",
        "done | split -l30\n",
        "\n",
        "cat gene*blastn | sort -u > geneflanks_allblastn #for later\n",
        "\n",
        "#Do the same with geneflanks:\n",
        "\n",
        "for i in $(cut -f 1 geneflanks_targets.txt)\n",
        "do\n",
        "  echo \"grep -P \\\"$i\\t\\\" geneflanks_targets.txt | sed \\\"s/,/\\\\n/g\\\" | sed \\\"s/\\t/\\\\n/g\\\" | sed \\\"s/^ *//g\\\" | tail -n+2 > ${i}_geneflanks_targetlist.txt\"\n",
        "  echo \"cat geneflanks_extragenus_blastn geneflanks_intragenus_blastn geneflanks_pangenome_blastn | grep -w -F -f ${i}_geneflanks_targetlist.txt - | grep \\\"${i}_\\\" | sed \\\"s/_up_/\\\\t/g\\\" | sed \\\"s/_down_/\\\\t/g\\\" | cut -f1,3- | awk -F '\\t' '{OFS=\\\"\\\"}{print \\$13,\\\"%\\\",\\$14,\\\"%\\\",\\$10,\\\"\\t\\\",\\$2}' | awk -F'\\t' '{ values[\\$2] = (values[\\$2] == \\\"\\\" ? \\$1 : values[\\$2] \\\", \\\" \\$1) } END { for (value in values) { print value \\\"\\t\\\" values[value] } }' | sed \\\"s/%plus, /%/g\\\" | sed \\\"s/%minus, /%/g\\\" | sed \\\"s/%plus/\\tplus/g\\\" | sed \\\"s/%minus/\\tminus/g\\\" | sed \\\"s/%/,/g\\\" | sed \\\"s/\\t/,/g\\\" | sed \\\"s/ //g\\\" | awk -F',' '{identifier = \\$1; values = \\$2 \\\",\\\" \\$3 \\\",\\\" \\$4 \\\",\\\" \\$5; split(values, array, \\\",\\\"); asort(array); middle1 = array[2]; middle2 = array[3]; difference = middle2 - middle1; if (difference >= 0) { print identifier, middle1, middle2, difference, \\$6; } else { print identifier, middle2, middle1, -difference, \\$6; } }' > ${i}_geneflanks_intervalinfo\"\n",
        "done | split -l30 - y\n",
        "\n",
        "#Now compile prox and gene flank information together in the same intervalinfo file:\n",
        "\n",
        "for i in $(ls *intervalinfo | rev | cut -f4- -d \"_\" | rev | sort -u)\n",
        "do\n",
        "ls \"$i\"_*intervalinfo | sed \"s/^/cat /g\" | bash >> \"$i\"_compiled_intervalinfo.txt\n",
        "done\n",
        "\n",
        "#Attach taxonomy info to each file\n",
        "#Make an interim file that contains the taxonomy information for all target contigs\n",
        "#Otherwise searching the massive database would take a long time\n",
        "cut -f1 -d \" \" *_compiled_intervalinfo.txt | sort -u > alltaxa.compiled_intervalinfo.txt\n",
        "cat /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.noescherichia.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_contig_taxa.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv | grep -w -F -f alltaxa.compiled_intervalinfo.txt - | cut -f2- > alltaxa.compiled_intervalinfo.interim\n",
        "sort -k1 alltaxa.compiled_intervalinfo.interim -o alltaxa.compiled_intervalinfo.interim\n",
        "\n",
        "#Attach taxonomy information from the nascent file to the interval info files\n",
        "for i in $(ls *_compiled_intervalinfo.txt | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "sort -k1 \"$i\"_compiled_intervalinfo.txt -o \"$i\"_compiled_intervalinfo.txt\n",
        "cut -f1 -d \" \" \"$i\"_compiled_intervalinfo.txt | sort -u > temp\n",
        "grep -w -F -f temp alltaxa.compiled_intervalinfo.interim | sort -k1 | join -1 1 -2 1 - \"$i\"_compiled_intervalinfo.txt > \"$i\"_compiled_intervalinfo.taxa.txt\n",
        "done\n",
        "\n",
        "#Use this code to remove unwanted interval infos——the genomes in which the gene is already present based on prior blasts\n",
        "#The code is also used to remove sequences from regular blastn stuff\n",
        "#My original code detected flanks in a conservative way——because the goal was to just detect non-coding orthologs\n",
        "#For the more liberal definition of flanks, see next \"debugging\" chunk\n",
        "\n",
        "for i in $(ls *_compiled_intervalinfo.taxa.txt | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "ortholog_name=$(echo $i | cut -f2 -d \"@\")\n",
        "genome_name=$(echo $i | cut -f1 -d \"@\")\n",
        "value=$(grep -w $ortholog_name ../../Ecoli_gtfs/$genome_name.filtered.nr.gtf | awk -F '\\t' '{print $5-$4+1}')\n",
        "sed \"s/$/ $value/\" \"$i\"_compiled_intervalinfo.taxa.txt > \"$i\"_compiled_intervalinfo.taxa.txt.2 #Append length\n",
        "echo \"$i\" | sed \"s/$/(/g\" | grep -F -f - ../all_genes_of_interest.presence.tsv | cut -f2 -d \" \" |\n",
        "grep -v -w -F -f - \"$i\"_compiled_intervalinfo.taxa.txt.2 |\n",
        "awk '(($5>($7*0.5))&&($5<10000))' > \"$i\"_compiled_intervalinfo.taxa.final.txt #note the filters\n",
        "done\n",
        "\n",
        "wc -l *final.txt | grep \" 0 \" | rev | cut -f1 -d \" \" | rev | sed \"s/^/rm /g\" | bash\n"
      ],
      "metadata": {
        "id": "VjfMoiT3tIPu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Debugging"
      ],
      "metadata": {
        "id": "lfs6AHGhaTsr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Let's make another set of files where flanks are defined more liberally\n",
        "\n",
        "mkdir flank_alternative_111025\n",
        "cd flank_alternative_111025\n",
        "cp ../*_compiled_intervalinfo.taxa.txt .\n",
        "\n",
        "for i in $(ls *_compiled_intervalinfo.taxa.txt | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "ortholog_name=$(echo $i | cut -f2 -d \"@\")\n",
        "genome_name=$(echo $i | cut -f1 -d \"@\")\n",
        "value=$(grep -w $ortholog_name ../../../Ecoli_gtfs/$genome_name.filtered.nr.gtf | awk -F '\\t' '{print $5-$4+1}')\n",
        "sed \"s/$/ $value/\" \"$i\"_compiled_intervalinfo.taxa.txt > \"$i\"_compiled_intervalinfo.taxa.txt.2\n",
        "echo \"$i\" | sed \"s/$/(/g\" | grep -F -f - ../../all_genes_of_interest.presence.tsv | cut -f2 -d \" \" | grep -v -w -F -f - \"$i\"_compiled_intervalinfo.taxa.txt.2 | awk '($5<($7*5))' > \"$i\"_compiled_intervalinfo.taxa.final.txt\n",
        "done\n",
        "\n",
        "wc -l *final.txt | grep \" 0 \" | rev | cut -f1 -d \" \" | rev | sed \"s/^/rm /g\" | bash"
      ],
      "metadata": {
        "id": "HGqBlsP1aT_i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Identifying homologs within flanking regions**"
      ],
      "metadata": {
        "id": "lAQEkruMdj_E"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#To detect homologs in flank-bounded regions, we extract +/-300bp from both edges of the interval\n",
        "for i in $(ls *_compiled_intervalinfo.taxa.final.txt | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "awk '{print $1,$2,$3-300,$4+300,$5,$6,$7}' \"$i\"_compiled_intervalinfo.taxa.final.txt | awk '{$3=($3<1?1:$3); $4=($4<1?1:$4); print}' > \"$i\"_compiled_intervalinfo.taxa.extended.txt\n",
        "done\n",
        "\n",
        "#We then use samtools to extract the sequences from the respective genomes\n",
        "#This is a time-consuming step, so it's placed within wrappers to parallelize\n",
        "#See Preparing_database.sh to see how the individual_genomes directory was made\n",
        "\n",
        "for variable in $(ls *_compiled_intervalinfo.taxa.extended.txt | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "    command=\"grep \\\"plus\\\" ${variable}_compiled_intervalinfo.taxa.extended.txt | awk '{OFS=\\\"\\\"}{print \\\"samtools faidx /stor/scratch/Ochman/hassan/100724_Complete_Genomes/individual_genomes/\\\",\\$1,\\\" \\\",\\$1,\\\":\\\",\\$3,\\\"-\\\",\\$4}' | sed \\\"s/\\$/ >> ${variable}_interval.faa/g\\\" && grep \\\"minus\\\" ${variable}_compiled_intervalinfo.taxa.extended.txt | awk '{OFS=\\\"\\\"}{print \\\"samtools faidx -i /stor/scratch/Ochman/hassan/100724_Complete_Genomes/individual_genomes/\\\",\\$1,\\\" \\\",\\$1,\\\":\\\",\\$3,\\\"-\\\",\\$4}' | sed \\\"s/\\$/ >> ${variable}_interval.faa/g\\\"\"\n",
        "    output_file=\"${variable}_joint_samtools.sh\"\n",
        "    bash -c \"$command\" > \"$output_file\"\n",
        "done\n",
        "\n",
        "ls *samtools.sh | sed \"s/^/bash /g\" > running.sh\n",
        "\n",
        "/stor/work/Ochman/hassan/tools/parallelize_run.samtools.sh\n",
        "\n",
        "###Sanity checks:\n",
        "#check for empty sequences:\n",
        "for i in *_interval.faa; do seqkit fx2tab $i | awk '(length($2)==0)'; done\n",
        "#none, good\n",
        "\n",
        "#check if interval sequences couldn't be extracted\n",
        "for i in $(ls *_interval.faa | rev | cut -f2- -d \"_\" | rev); do grep \"^>\" \"$i\"_interval.faa | wc -l; wc -l \"$i\"_joint_samtools.sh; done | paste - - | awk '($1!=$2)'\n",
        "#All were extractable\n",
        "\n",
        "#Now turn the extracted sequences into blast databases:\n",
        "for i in $(ls *_interval.faa | rev | cut -f2- -d \"_\" | rev)\n",
        "do\n",
        "makeblastdb -in \"$i\"_interval.faa -dbtype nucl -out \"$i\"_interval\n",
        "done\n",
        "\n",
        "#We then blastn the gene sequences against the extracted sequences:\n",
        "#Since the databases are small(er)——only sequences bounded by flanks——we can get away with using a very sensitive search parameter, -word 7\n",
        "for i in $(ls *_interval.faa | rev | cut -f2- -d \"_\" | rev)\n",
        "do\n",
        "ls \"$i\"_interval.faa | rev | cut -f2- -d \"_\" | rev | sed \"s/$/(/g\" | grep --no-group-separator -A1 -f - ../all_genes_of_interest.CDS.faa |\n",
        "blastn -query - -db \"$i\"_interval -outfmt 0 -num_threads 72 -num_descriptions 13638 -num_alignments 13638 -evalue 200000 -out \"$i\"_interval_blastn -word_size 7\n",
        "done\n",
        "\n",
        "#Now we convert the blast output to a parseable mview file\n",
        "export PERL5LIB=/stor/scratch/Ochman/hassan/genomics_toolbox/mview-1.67/lib/\n",
        "\n",
        "#Since mview takes a while, we put the code in \"echo\" to parallelize it\n",
        "for i in $(ls *_interval_blastn | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "echo \"/stor/scratch/Ochman/hassan/genomics_toolbox/mview-1.67/bin/mview -in blast ${i}_interval_blastn > ${i}_blastn_mviewed\"\n",
        "done > running.sh\n",
        "\n",
        "/stor/work/Ochman/hassan/tools/parallelize_run.sh #helper script used for parallelization\n",
        "\n",
        "#We now parse the mviewed file according to some criteria\n",
        "#We get rid of any case where the total combined length of all alignments doesn't cover 50% of query length\n",
        "#And also targets whose alignments don't cover the query by at least 50%\n",
        "#Since there's no e-value cutoff, it's necessary to utilize a query coverage cutoff to get rid of spurious hits\n",
        "#This might miss out on very small regions of homologous non-coding sequences\n",
        "#But that's fine, since we're independently collecting data on which of these genes had flanks at reasonable distances\n",
        "for i in $(ls *mviewed | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "querylength=$(echo $i | sed \"s/$/(/g\" | grep -F -A1 -f - ../all_genes_of_interest.CDS.faa | grep -v \"^>\" | awk '{print length($0)}')\n",
        "ratio=$(tail -n+8 \"$i\"_blastn_mviewed | head -1 | awk '{print $(NF-1)}' | sed \"s/:/\\t/g\" | awk -v var=$querylength -F '\\t' '{print ($2-$1+1)/var}')\n",
        "if (( $(echo \"$ratio > 0.5\" | bc -l) ))\n",
        "then\n",
        "tail -n+9 \"$i\"_blastn_mviewed | head -n-3 | awk '{print $2,$(NF-4),$NF}' | sed \"s/%//g\" | awk '($2>50)' | awk '{print $1,$3}' | sed \"s/^/>/g\" | sed \"s/ /\\n/g\" > \"$i\"_blastn_seq.faa\n",
        "fi\n",
        "done\n",
        "\n",
        "wc -l *_blastn_seq.faa | grep \" 0 \" | rev | cut -f1 -d \" \" | rev | sed \"s/^/rm /g\" | bash\n",
        "\n",
        "#Adding taxonomic info back exactly as I did above:\n",
        "\n",
        "#Attach taxonomy info to each file\n",
        "#make an interim file containing contig names from all these files\n",
        "cat *_blastn_seq.faa | grep \"^>\" | tr -d \">\" | rev | cut -f2- -d \":\" | rev | sort -u > alltaxa.compiled_intervalinfo.txt\n",
        "cat /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.noescherichia.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_contig_taxa.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv | grep -w -F -f alltaxa.compiled_intervalinfo.txt - | cut -f2- > alltaxa.compiled_intervalinfo.interim\n",
        "sort -k1 alltaxa.compiled_intervalinfo.interim -o alltaxa.compiled_intervalinfo.interim\n",
        "\n",
        "#Extract taxonomy info with this code\n",
        "for i in $(ls *_blastn_seq.faa | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "echo \"seqkit fx2tab ${i}_blastn_seq.faa | sed \\\"s/\\t$//g\\\" | awk -F \\\":\\\" '{print \\$1\\\"\\t\\\"\\$0}' | sort -k1 > ${i}_blastn_seq.tsv\"\n",
        "echo \"cut -f1 ${i}_blastn_seq.tsv | sort -u > ${i}.temp\"\n",
        "echo \"grep -w -F -f ${i}.temp alltaxa.compiled_intervalinfo.interim | sort -k1 | join -1 1 -2 1 - ${i}_blastn_seq.tsv > ${i}_blastn_seq.interim.tsv\"\n",
        "done\n",
        "\n",
        "for i in $(ls *_blastn_seq.interim.tsv | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "awk '{print \">flank_\"$2\"_\"$3,$4}' \"$i\"_blastn_seq.interim.tsv | sed \"s/ /\\n/g\" > \"$i\"_blastn.seq.flank.faa #Mark the sequences with \"flank\"\n",
        "done"
      ],
      "metadata": {
        "id": "pasyDSVeYyvQ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}