{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "In the section below, we pool the sequences against which our gene had a hit -- both from the flank-guided and flank-unguided approaches, we extract all open reading frames from them (same strand), and see if our query protein has a hit against any of those proteins."
      ],
      "metadata": {
        "id": "PLB1_m4Yru3G"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RVtZorT7qP4W"
      },
      "outputs": [],
      "source": [
        "#First, let's compile both sets of sequences together\n",
        "\n",
        "mkdir ORFpresence\n",
        "cd ORFpresence\n",
        "#Make a list of query proteins for which we have either flank-guided or unguided DNA sequence hits\n",
        "ls ../flank/*_blastn.seq.*.faa ../regular/*_blastn.seq.*.faa | rev | cut -f1 -d \"/\" | cut -f2- -d \"_\" |  rev | sort -u > ORFsearch_candidates.txt\n",
        "\n",
        "for i in $(cat ORFsearch_candidates.txt)\n",
        "do\n",
        "cat ../flank/\"$i\"_blastn.seq.flank.faa ../regular/\"$i\"_blastn.seq.regular.faa >> \"$i\"_compiled_redundant.faa #Compile the sequences together\n",
        "done\n",
        "\n",
        "#Reduce in size by remove redundant sequences\n",
        "#I.e., if identical sequences exist for the same taxon of relevance (genomes for intra-pangenome, species or genus names outside)——only retain one\n",
        "\n",
        "for i in $(ls *_compiled_redundant.faa | rev | cut -f3- -d \"_\" | rev)\n",
        "do\n",
        "cat \"$i\"_compiled_redundant.faa | seqkit fx2tab | sed \"s/Escherichia_/E/g\" | grep \"Ecoli@\" | sed \"s/_/@/1\" | awk -F \"@\" '{print $2\"@\"$3\"\\t\"$0}' | awk '{print $2\"\\t\"$1\"%\"$NF}' | awk '!seen[$2]++' | sed \"s/%/ /g\" | awk '{print \">\"$2\"%\"$1\" \"$3}' | sed \"s/\\t$//g\" | sed \"s/ /\\n/g\" > \"$i\".ORFsearch_targets.faa\n",
        "cat \"$i\"_compiled_redundant.faa | seqkit fx2tab | sed \"s/Escherichia_/E/g\" | grep -v \"Ecoli@\" | awk -F \"_\" '{print $2,$0}' | awk '{print $2,$1\"%\"$3}' | awk '!seen[$2]++' | sed \"s/%/ /g\" | awk '{print \">\"$2\"%\"$1\" \"$3}' | sed \"s/\\t$//g\" | sed \"s/ /\\n/g\" >> \"$i\".ORFsearch_targets.faa\n",
        "done\n",
        "\n",
        "#Now to extract ORFs from these sequences and search the query against them\n",
        "\n",
        "for i in $(ls ../*.ORFsearch_targets.faa | cut -f2- -d \"/\" | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "#Print the sequence in three different frames (same strand) for downstream processing\n",
        "seqkit fx2tab ../\"$i\".ORFsearch_targets.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" | awk '/^>/ {print} !/^>/ {gsub(/-/, \"\"); print}' | head -n-2 > \"$i\".ORFsearch_targets.frames.faa\n",
        "seqkit fx2tab ../\"$i\".ORFsearch_targets.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" | awk '/^>/ {print} !/^>/ {gsub(/-/, \"\"); print}' | head -n-2 | awk '/^>/ {print} !/^>/ {print substr($0, 2)}' | sed \"s/>/>frame1_/g\" >> \"$i\".ORFsearch_targets.frames.faa\n",
        "seqkit fx2tab ../\"$i\".ORFsearch_targets.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" | awk '/^>/ {print} !/^>/ {gsub(/-/, \"\"); print}' | head -n-2 | awk '/^>/ {print} !/^>/ {print substr($0, 3)}' | sed \"s/>/>frame2_/g\" >> \"$i\".ORFsearch_targets.frames.faa\n",
        "#Now identify all open reading frames using this custom code\n",
        "awk '/^>/ {print} !/^>/ {gsub(/.{3}/, \"& \"); print}' \"$i\".ORFsearch_targets.frames.faa | sed \"s/ATG/M/g\" | sed \"s/TTG/L/g\" | sed \"s/GTG/V/g\" | sed \"s/TAA/Z/g\" | sed \"s/TGA/Z/g\" | sed \"s/TAG/Z/g\" | sed \"s/ //g\" | awk '/^>/ {print} !/^>/ {print $0 \"Z\"}' | awk '/^>/ {header = $0} !/^>/ {\n",
        "    while(match($0, /[MLV][^Z]*Z/)) {\n",
        "        orf = substr($0, RSTART, RLENGTH)\n",
        "        print header \": \" orf\n",
        "        $0 = substr($0, RSTART + RLENGTH)  # Remove found ORF and continue searching\n",
        "    }\n",
        "}' | cat -n | sed \"s/^ *//g\" | sed \"s/\\t>/_ORF_/g\" | sed \"s/^/>/g\" | sed \"s/: /\\n/g\" |\n",
        "awk '/^>/ {print} !/^>/ {gsub(\"M\", \"ATG\"); gsub(\"L\", \"TTG\"); gsub(\"V\", \"GTG\"); print}' > \"$i\".ORFsearch_targets.ORFs.faa\n",
        "#Translate the ORFs into proteins\n",
        "/stor/work/Ochman/hassan/tools/faTrans \"$i\".ORFsearch_targets.ORFs.faa \"$i\".ORFsearch_targets.ORFs.prot.faa #Extract all ORFs from the +strand of each genome\n",
        "#Get the query sequence and translate it\n",
        "echo $i | sed \"s/$/(/g\" | grep --no-group-separator -A1 -f - ../../all_genes_of_interest.CDS.faa > temp\n",
        "/stor/work/Ochman/hassan/tools/faTrans -stop temp \"$i\".query.prot.faa #Query gene\n",
        "#Finally, run the fasta36 (ssearch) search\n",
        "/stor/work/Ochman/hassan/tools/fasta-36.3.8i/bin/ssearch36 -m 8 \"$i\".query.prot.faa \"$i\".ORFsearch_targets.ORFs.prot.faa > \"$i\".results.ssearch.tab\n",
        "done\n",
        "\n",
        "#Process the results to get a list of all genomes (or taxa for extra-pangenome) which has a matching ORF, either full or partial\n",
        "for i in $(ls ../*.ORFsearch_targets.faa | cut -f2- -d \"/\" | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "length=$(seqkit fx2tab \"$i\".query.prot.faa | awk '{print length($2)}')\n",
        "awk -F '\\t' -v var=\"$length\" '{cov = ($8 - $7 + 1) / var; print $0 \"\\t\" cov}' \"$i\".results.ssearch.tab | awk -F '\\t' '($NF>0.59999&&$11<0.001)' | cut -f2 | sed \"s/frame1_//g\" | sed \"s/frame2_//g\" | cut -f3- -d \"_\" | awk '($1~\"@\")' | cut -f1 -d \"%\" | sed \"s/^/\"$i\"\\t/g\" >> ORF_presence.eval001.tsv\n",
        "awk -F '\\t' -v var=\"$length\" '{cov = ($8 - $7 + 1) / var; print $0 \"\\t\" cov}' \"$i\".results.ssearch.tab | awk -F '\\t' '($NF>0.59999&&$11<0.001)' | cut -f2 | sed \"s/frame1_//g\" | sed \"s/frame2_//g\" | cut -f3- -d \"_\" | awk '($1!~\"@\")' | cut -f1 -d \"%\" | sed \"s/^/\"$i\"\\t/g\" >> ORF_presence.eval001.tsv\n",
        "awk -F '\\t' -v var=\"$length\" '{cov = ($8 - $7 + 1) / var; print $0 \"\\t\" cov}' \"$i\".results.ssearch.tab | awk -F '\\t' '($NF<0.6&&$11<0.001)' | cut -f2 | sed \"s/frame1_//g\" | sed \"s/frame2_//g\" | cut -f3- -d \"_\" | awk '($1~\"@\")' | cut -f1 -d \"%\" | sed \"s/^/\"$i\"\\t/g\" >> ORF_partial.eval001.tsv\n",
        "awk -F '\\t' -v var=\"$length\" '{cov = ($8 - $7 + 1) / var; print $0 \"\\t\" cov}' \"$i\".results.ssearch.tab | awk -F '\\t' '($NF<0.6&&$11<0.001)' | cut -f2 | sed \"s/frame1_//g\" | sed \"s/frame2_//g\" | cut -f3- -d \"_\" | awk '($1!~\"@\")' | cut -f1 -d \"%\" | sed \"s/^/\"$i\"\\t/g\" >> ORF_partial.eval001.tsv\n",
        "done\n",
        "\n",
        "#These files will be required eventually for purposes of compiling results"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "I need to do one more search to rule out ORFans: Against nr.dmnd.\n",
        "\n",
        "I don't trust this database——it contains contaminated or otherwise unreliable records. So I will only use this to rule out false positives only when other approaches have been implemented."
      ],
      "metadata": {
        "id": "mf3I_6ZvBhLG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond blastp \\\n",
        "  -q step2_genusspecific_ORFan.replaced.faa \\\n",
        "  -d /stor/scratch/Ochman/hassan/nr/03272024_nr.dmnd \\\n",
        "  --taxonmap /stor/scratch/Ochman/hassan/nr/prot.accession2taxid.FULL \\\n",
        "  --taxonnames /stor/scratch/Ochman/hassan/nr/names.dmp \\\n",
        "  --taxonnodes /stor/scratch/Ochman/hassan/nr/nodes.dmp \\\n",
        "  --outfmt 6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps qstart qend sstart send qlen slen evalue bitscore staxids sscinames \\\n",
        "  --ultra-sensitive \\\n",
        "  --out step2genus_vs_nr_withtax.tsv \\\n",
        "  -k 0 -b8 -c1\n",
        "\n",
        "cat step2genus_vs_nr_withtax.tsv |\n",
        "awk -F '\\t' '($5>60&&$16<0.001)' |\n",
        "awk -F '\\t' '($19!~\"Escherichia coli\")' |\n",
        "awk -F '\\t' '($19!~\"Shigella\")' |\n",
        "awk -F '\\t' '($19!~\"Escherichia sp.\")' |\n",
        "awk -F '\\t' '($19!=\"N/A\")' |\n",
        "awk -F '\\t' '($19!~\"Enterobacteriaceae\")' |\n",
        "awk -F '\\t' '($19!=\"Escherichia\")' |\n",
        "awk -F '\\t' '($19!=\"Caudoviricetes sp.\")' |\n",
        "awk -F '\\t' '($19!=\"unclassified Escherichia\")' |\n",
        "awk -F '\\t' '($19!=\"Bacteriophage sp.\")' |\n",
        "awk -F '\\t' '($19!=\"uncultured bacterium\")' |\n",
        "awk -F '\\t' '($19!=\"Klebsiella pneumoniae IS22\")' |\n",
        "awk -F '\\t' '($19!=\"Enterobacterales\")' |\n",
        "awk -F '\\t' '($19!=\"Escherichia;uncultured bacterium\")' |\n",
        "awk -F '\\t' '($19!=\"Enterobacter sp. EC-NT1\")' |\n",
        "awk -F '\\t' '($19!=\"Escherichia;Caudoviricetes sp.\")' |\n",
        "awk -F '\\t' '($19!=\"Salmonella sp. S13\")' |\n",
        "awk -F '\\t' '($19!=\"Escherichia;Salmonella sp. S13\")' |\n",
        "awk -F '\\t' '($19!~\"phage\")' |\n",
        "cut -f1,19 | sort -u | cut -f1 | sort | uniq -c | awk '($1>1)' | rev |\n",
        "cut -f1 -d \" \" | rev | cut -f1 -d \"(\" | grep -v -F -f - step2_genusspecific_ORFan.replaced.txt > step3_genusspecific_ORFan.replaced.txt"
      ],
      "metadata": {
        "id": "SaVo4RIrBgd9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# @title\n",
        "#sort -k1 ORF_presence.tsv -o ORF_presence.tsv\n",
        "#for i in $(ls ../*.mafft_input.faa | cut -f2- -d \"/\" | rev | cut -f3- -d \".\" | rev)\n",
        "#do\n",
        "#echo -n $i \" \" && grep \"$i\" ../\"$i\".mafft_input.faa | tr -d \">\"\n",
        "#done | sort -k1 | join -1 1 -2 1 ORF_presence.tsv - | awk '{print $3\"\\t\"$2}' > ORF_approach_presence.tsv #I guess this was required to fix differences in query name\n",
        "\n",
        "#Exclude non-ORFans from OG list:\n",
        "egrep -v \"Ecoli|Ealbertii|Efergu|Eruysi|Ewhit|Emarmot\" alignments/ORF_approach/ORF_approach_presence.tsv | grep -v \"Enterobacteriaceae\" | cut -f1 -d \"(\" | sort -u | sed \"s/$/(/g\" | grep -v -F -f - step1_genusspecific_ORFan.replaced.txt | cut -f1 -d \"(\" | sort -u > step2_genusspecific_ORFan.replaced.txt\n",
        "\n",
        "#nr blast:\n",
        "\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond blastp -q step1_genusspecific_ORFans.faa -d /stor/scratch/Ochman/hassan/nr/03272024_nr.dmnd --outfmt 6 qseqid sseqid pident nident qcovhsp length mismatch gapopen gaps qstart qend sstart send qlen slen evalue bitscore --ultra-sensitive --out all_proteins_vs_GBRS_annotated.tsv -k 0 -b8 -c1\n",
        "\n",
        "for i in $(ls ../*.mafft_input.faa | cut -f2- -d \"/\" | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "#/stor/work/Ochman/hassan/tools/fasta-36.3.8i/bin/fasta36 -m 9c \"$i\".query.prot.faa \"$i\".mafft_input.searchtarget.prot.faa > \"$i\".results.2.tab\n",
        "length=$(seqkit fx2tab \"$i\".query.prot.faa | awk '{print length($2)}')\n",
        "awk '($22~\"=\")' \"$i\".results.2.tab | awk '{print $1,$6,$7,$11,$12,$19}' | awk -v var=\"$length\" '{cov = ($5 - $4 - $NF + 1) / var; print $0\" \"cov}' | awk '($2<0.001&&$NF>0.6)' | sed \"s/Escherichia_/E/g\" | sed \"s/flank_/flank\\t/g\" | sed \"s/regular_/regular\\t/g\"\n",
        "done"
      ],
      "metadata": {
        "cellView": "form",
        "id": "4UrHuJDqAXKb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Compiling all data together"
      ],
      "metadata": {
        "id": "vOzIOCrPoKIr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Files that reports presence:\n",
        "\n",
        "cat /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/all_genes_of_interest.presence.tsv /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/alignments/ORF_approach/ORF_approach_presence.tsv | sed \"s/Escherichia_/E/g\" | sed \"s/(+)//g\" | sed \"s/(-)//g\" | sort -u > presence.tsv\n",
        "\n",
        "#Files that report putative non-coding homolog:\n",
        "grep \"^>\" alignments/*mafft_input.faa | cut -f2- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/_flank/\\t/g\" | sed \"s/_regular/\\t/g\" | grep -v \")$\" | cut -f1,2 | sed \"s/.mafft_input.faa\\t>/\\t/g\" | sort -u > putative_noncoding.tsv\n",
        "\n",
        "#Files that report presence of flanks:\n",
        "grep \" \" flank/flank_alternative_111025/*final.txt | cut -f3- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/_compiled_intervalinfo.taxa.final.txt//g\" | sed \"s/ /\\t/g\" | cut -f1,3 | sed \"s/Escherichia_/E/g\" > flank_presence.tsv\n",
        "\n",
        "for i in $(cat step1_genusspecific_ORFan.replaced.txt | cut -f1 -d \"(\"); do\n",
        "  for j in $(cat alignments/all_lineages_ordered.txt); do\n",
        "\n",
        "    if awk -v a=\"$i\" -v b=\"$j\" '$1==a && $2==b' presence.tsv | grep -q .; then\n",
        "      echo -e \"$i\\t$j\\tpresent\"\n",
        "\n",
        "    elif awk -v a=\"$i\" -v b=\"$j\" '$1==a && $2==b' putative_noncoding.tsv | grep -q .; then\n",
        "      echo -e \"$i\\t$j\\tnoncoding\"\n",
        "\n",
        "    elif awk -v a=\"$i\" -v b=\"$j\" '$1==a && $2==b' flank_presence.tsv | grep -q .; then\n",
        "      echo -e \"$i\\t$j\\tflanks\"\n",
        "\n",
        "    else\n",
        "      echo -e \"$i\\t$j\\tnontraceable\"\n",
        "    fi\n",
        "\n",
        "  done\n",
        "done\n",
        "\n",
        "for i in $(cut -f1 -d \"(\" step1_genusspecific_ORFan.replaced.txt | head -1); do\n",
        "  echo 'for j in $(cat alignments/all_lineages_ordered.txt); do'\n",
        "  echo \"  if awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$2==b' presence.tsv | grep -q .; then\"\n",
        "  echo \"    echo -e \\\"$i\\t\\$j\\tpresent\\\"\"\n",
        "  echo \"  elif awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$2==b' putative_noncoding.tsv | grep -q .; then\"\n",
        "  echo \"    echo -e \\\"$i\\t\\$j\\tnoncoding\\\"\"\n",
        "  echo \"  elif awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$2==b' flank_presence.tsv | grep -q .; then\"\n",
        "  echo \"    echo -e \\\"$i\\t\\$j\\tflanks\\\"\"\n",
        "  echo \"  else\"\n",
        "  echo \"    echo -e \\\"$i\\t\\$j\\tnontraceable\\\"\"\n",
        "  echo \"  fi\"\n",
        "  echo 'done >> \\\"$i\\\".matrix.input.tsv'\n",
        "done"
      ],
      "metadata": {
        "id": "6YvTaeToW-8b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Compilation**"
      ],
      "metadata": {
        "id": "7xX7PHw9W3C_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "First, compile all genomes where the gene is present:"
      ],
      "metadata": {
        "id": "z2YY1dUxLmAb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#From protein blasts, pangenome:\n",
        "sort -k2 all_genes_of_interest.pangenome.protein.presenceabsence.tsv | join -1 2 -2 2 - /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_protein_taxa.tsv | awk '{OFS=\"\\t\"}{print $2,$NF,$3}' | sort -u > presence.tsv\n",
        "sort -k2 all_genes_of_interest.pangenome.contig.presenceabsence.tsv | join -1 2 -2 2 - /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv | awk '{OFS=\"\\t\"}{print $2,$NF,$3}' | sort -u >> presence.tsv\n",
        "\n",
        "#From protein blasts, non-pangenome:\n",
        "cat /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/all_genes_of_interest.presence.tsv | awk '($2!~\"@\")' | sort -u | sed \"s/Escherichia_/E/g\" >> presence.tsv\n",
        "\n",
        "#From ORF search:\n",
        "awk -F '\\t' '($2~\"@\")' /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/alignments/ORF_approach_111725/ORF_presence.eval001.tsv | sed -E 's/(Ecoli@[0-9]+)_/\\1\\t/g' | awk 'NR==FNR{map[$1]=$1;next}{for(k in map) if(index(k,$3)){ $3=map[k];break } print}' 450_lineage_designations.tsv - | sort -u >> presence.tsv\n",
        "awk -F '\\t' '($2!~\"@\")' /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/alignments/ORF_approach_111725/ORF_presence.eval001.tsv | sort -u >> presence.tsv\n",
        "\n",
        "#Massage:\n",
        "sed \"s/(+)//g\" presence.tsv | sed \"s/(-)//g\" | sort -u | sed \"s/Escherichia_/E/g\" | sed \"s/ /\\t/g\" > temp && mv temp presence.tsv"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 216
        },
        "collapsed": true,
        "id": "cYuvX-uFXHTO",
        "outputId": "695c6b31-3e21-435a-9461-876c1071b7b5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'all_genes_of_interest' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3522741335.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m#Present:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mall_genes_of_interest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextragenus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpresenceabsence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtsv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0mall_genes_of_interest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintragenus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpresenceabsence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtsv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mall_genes_of_interest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpangenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpresenceabsence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtsv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'all_genes_of_interest' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, compile all gnomes where the gene is partially present/has a noncoding status:"
      ],
      "metadata": {
        "id": "SW1kXvaKLp37"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Every case where blastn following synteny guided or unguided status had a hit:\n",
        "grep \"^>\" alignments/*_compiled_redundant.faa | cut -f2- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/>flank_//g\" | sed \"s/>regular_//g\" | sed \"s/Escherichia_/E/g\" | grep \"Ecoli@\" | cut -f-3 -d \"@\" | sed \"s/_compiled_redundant.faa//g\" | sed -E 's/(Ecoli@[0-9]+)_/\\1\\t/g' >> putative_noncoding.tsv\n",
        "grep \"^>\" alignments/*_compiled_redundant.faa | cut -f2- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/>flank_//g\" | sed \"s/>regular_//g\" | sed \"s/Escherichia_/E/g\" | grep -v \"Ecoli@\" | sed \"s/_compiled_redundant.faa//g\" | awk '{ split($2, a, \"_\"); print $1, a[1] }' | cut -f2 | sort -u | sed \"s/ /\\t/g\" >> putative_noncoding.tsv\n",
        "\n",
        "#All the partial hits:\n",
        "cat all_genes_of_interest.presencepartial.tsv | sort -u | sed \"s/ /\\t/g\" >> putative_noncoding.tsv\n",
        "\n",
        "#Massage:\n",
        "sed \"s/(+)//g\" putative_noncoding.tsv | sed \"s/(-)//g\" | sort -u | sed \"s/Escherichia_/E/g\" | sed \"s/ /\\t/g\" > temp && mv temp putative_noncoding.tsv\n",
        "\n",
        "#Flanks:\n",
        "grep \" \" flank/flank_alternative_111025/*final.txt | cut -f3- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/_compiled_intervalinfo.taxa.final.txt//g\" | sed \"s/ /\\t/g\" | awk -F '\\t' '($3~\"@\")' | awk -F '\\t' '{print $1,$3,$2}' | rev | cut -f2- -d \"@\" | rev | sed \"s/ /\\t/g\" | sort -u > flank_presence.tsv\n",
        "grep \" \" flank/flank_alternative_111025/*final.txt | cut -f3- -d \"/\" | sed \"s/:/\\t/1\" | sed \"s/_compiled_intervalinfo.taxa.final.txt//g\" | sed \"s/ /\\t/g\" | awk -F '\\t' '($3!~\"@\")' | cut -f1,3 | sed \"s/Escherichia_/E/g\" >> flank_presence.tsv"
      ],
      "metadata": {
        "id": "5csvLWS-LtR2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "PREPARE THE FIGURE"
      ],
      "metadata": {
        "id": "VcwnHu_FQl-P"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#To generate the figure, we only focus on a reduced set of genes that are species-specific.\n",
        "\n",
        "#To this end, we collapse the gene families further using a 60/60 cutoff\n",
        "#Let's do an all-vs-all blast, followed by Silix-based clustering:\n",
        "\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in step3_genusspecific_ORFan.replaced.faa --db step3_genusspecific_ORFan.replaced\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond blastp -q step3_genusspecific_ORFan.replaced.faa -d step3_genusspecific_ORFan.replaced --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore --ultra-sensitive --out step3_genusspecific_ORFan.replaced.silixinput.tsv -k 0 -b8 -c1\n",
        "\n",
        "#Silix clustering:\n",
        "/stor/work/Ochman/hassan/tools/silix-1.3.0/src/silix -f cluster_ -i 0.6 -r 0.6 -q -2 -s 3 step3_genusspecific_ORFan.replaced.faa step3_genusspecific_ORFan.replaced.silixinput.tsv | sort -k2 > silix_output.step3.tsv\n",
        "\n",
        "#Extract the longest gene from each cluster to be the representative:\n",
        "seqkit fx2tab step3_genusspecific_ORFan.replaced.faa | awk '{print $1,length($2)}' | sort -k1 | join -1 1 -2 2 - silix_output.step3.tsv | sort -k3,3 -k2,2nr | awk '!seen[$3]++' | cut -f1 -d \" \" | grep --no-group-separator -A1 -F -f - step3_genusspecific_ORFan.replaced.faa > step3_genusspecific_ORFan.replaced.reduced.faa\n",
        "\n",
        "#Only retain species-specific genes:\n",
        "awk '($2!~\"@\")' presence.tsv | grep -F -f <(grep \"^>\" step3_genusspecific_ORFan.replaced.reduced.faa | tr -d \">\" | cut -f1 -d \"(\") - | cut -f1 | sort -u | grep -v -F -f - step3_genusspecific_ORFan.replaced.reduced.faa  | grep \"^>\" | grep --no-group-separator -A1 -F -f - step3_genusspecific_ORFan.replaced.reduced.faa > step3_speciesspecific_ORFan.replaced.reduced.faa\n",
        "\n",
        "for i in $(grep \"^>\" step3_speciesspecific_ORFan.replaced.reduced.faa | tr -d \">\" | cut -f1 -d \"(\"); do\n",
        "  echo 'for j in $(cut -f1 -d \" \" ../450_lineage_designations.tsv); do'\n",
        "  echo '  lineage=$(awk -v key=\"$j\" '\"'\"'$1==key {print $2}'\"'\"' ../450_lineage_designations.tsv)'\n",
        "  echo \"  if awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$3==b' presence.tsv | grep -q .; then\"\n",
        "  echo '    echo -e \"$j\\t$lineage\\tpresent\"'\n",
        "  echo \"  elif awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$3==b' putative_noncoding.tsv | grep -q .; then\"\n",
        "  echo '    echo -e \"$j\\t$lineage\\tnoncoding\"'\n",
        "  echo \"  elif awk -v a=\\\"$i\\\" -v b=\\\"\\$j\\\" '\\$1==a && \\$3==b' flank_presence.tsv | grep -q .; then\"\n",
        "  echo '    echo -e \"$j\\t$lineage\\tflanks\"'\n",
        "  echo \"  else\"\n",
        "  echo '    echo -e \"$j\\t$lineage\\tabsent\"'\n",
        "  echo \"  fi\"\n",
        "  echo \"done >> \\\"${i}.matrix.input.tsv\\\"\"\n",
        "done\n",
        "\n",
        "cp ../species_specific_ORFans_deepdive/step1_lineage.tsv .\n",
        "\n",
        "#CONTINUE FROM HERE:\n",
        "\n",
        "sed -i \"s/%/\\t/g\" *matrix.input.tsv\n",
        "\n",
        "for i in $(ls *matrix.input.tsv | rev | cut -f4- -d \".\" | rev)\n",
        "do\n",
        "awk -F'\\t' 'NR==FNR { order[$2]=$1; next } { key=($3 in order)?order[$3]:999; printf \"%03d\\t%s\\n\", key, $0 }' step1_lineage.tsv \"$i\".matrix.input.tsv > \"$i\".step2_prefixed.tsv\n",
        "sort -k1,1n -k2 \"$i\".step2_prefixed.tsv | cut -f2- | cut -f3- | awk '{ if (!seen[$1]++) order[++n]=$1; vals[$1]=vals[$1]?vals[$1]\",\"$2:$2 } END { for(i=1;i<=n;i++){k=order[i]; printf \"%s(%s)%s\", k, vals[k], (i<n?\",\":\"\\n\") } }' | sed \"s/^/\"$i\",/g\" >> species_specific_distributions.allgenomes.transformed.csv\n",
        "done\n",
        "\n",
        "cat species_specific_distributions.allgenomes.transformed.csv | sed \"s/,/\\t/1\" | sed \"s/),/)\\t/g\" | sed \"s/,/%/g\" | sed \"s/\\t/,/g\" |\n",
        "awk -F ',' '\n",
        "{\n",
        "  printf \"%s\", $1\n",
        "  for (i = 2; i <= NF; i++) {\n",
        "\n",
        "    match($i, /^[0-9]+/)\n",
        "    num = substr($i, RSTART, RLENGTH)\n",
        "\n",
        "    start = index($i, \"(\")\n",
        "    contents = substr($i, start + 1, length($i) - start - 1)\n",
        "\n",
        "    # reset counters\n",
        "    n_present = n_noncoding = n_flanks = 0\n",
        "\n",
        "    # count occurrences\n",
        "    n_present    = gsub(/present/,    \"\", contents)\n",
        "    n_noncoding  = gsub(/noncoding/,  \"\", contents)\n",
        "    n_flanks     = gsub(/flanks/,     \"\", contents)\n",
        "\n",
        "    # special handling for num == 14\n",
        "    if (num == \"14\") {\n",
        "\n",
        "      # divide by 2 (rounding up)\n",
        "      n_present   = int((n_present   + 1) / 2)\n",
        "      n_noncoding = int((n_noncoding + 1) / 2)\n",
        "      n_flanks    = int((n_flanks    + 1) / 2)\n",
        "\n",
        "      # if total > 10, subtract 1 from noncoding\n",
        "      if (n_present + n_noncoding + n_flanks > 10) {\n",
        "        n_noncoding--\n",
        "        if (n_noncoding < 0) n_noncoding = 0\n",
        "      }\n",
        "    }\n",
        "\n",
        "    # print 3-value tuple\n",
        "    printf \",%s(%d,%d,%d)\", num, n_present, n_noncoding, n_flanks\n",
        "  }\n",
        "  print \"\"\n",
        "}' > species_specific_distributions.allgenomes.transformed.elegantcode.1.csv\n",
        "\n",
        "awk -F'[()]' '{sum=0; for(i=2;i<=NF;i+=2){split($i,f,\",\"); sum+=f[1]} print $0 \"\\t\" sum}' species_specific_distributions.allgenomes.transformed.elegantcode.1.csv | awk -F'[()]' '{sum=0; for(i=2;i<=NF;i+=2){split($i,f,\",\"); sum+=f[2]} print $0 \"\\t\" sum}' | awk '{print $0,$2+$3}' | sort -nrk2 > species_specific_distributions.allgenomes.transformed.elegantcode.2.geneprofilemapping.csv\n",
        "\n",
        "cut -f2- -d \",\" species_specific_distributions.allgenomes.transformed.elegantcode.2.geneprofilemapping.csv | sort -u | sort -nrk2 | cut -f1 |\n",
        "awk '{out=\"\"; while(match($0, /\\(([0-9]+),/, m)){ out = (out==\"\" ? m[1] : out\",\"m[1]); $0 = substr($0, RSTART+RLENGTH) } print out }' > allgreencounts.csv\n",
        "\n",
        "cut -f2- -d \",\" species_specific_distributions.allgenomes.transformed.elegantcode.2.geneprofilemapping.csv | sort -u | sort -nrk2 | cut -f1 |\n",
        "awk '{out=\"\"; while(match($0, /,([0-9]+),/, m)){ out=(out==\"\"?m[1]:out\",\"m[1]); $0=substr($0,RSTART+RLENGTH)} print out }' > allredcounts.csv\n",
        "\n",
        "cut -f2- -d \",\" species_specific_distributions.allgenomes.transformed.elegantcode.2.geneprofilemapping.csv | sort -u | sort -nrk2 | cut -f1 |\n",
        "awk '{out=\"\"; while(match($0, /,([0-9]+)\\)/, m)){ out=(out==\"\"?m[1]:out\",\"m[1]); $0=substr($0,RSTART+RLENGTH)} print out }' > allyellowcounts.csv"
      ],
      "metadata": {
        "id": "k3pEVBm0Qodt"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}