{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# **Preparing databases**\n",
        "\n",
        "The heart of this pipeline involves mapping the presence and absence of genes across a large number of genomes: to other E. coli genomes (within-pangenome), to non-E. coli genomes that are nonetheless very closely related (Escherichia genomes excluding E. coli), and bacterial genomes in general (genomes other than those of the Esherichia genus). This involves making databases out of genomes and proteins from these groups, against which our genes of interest are to be subsequently searched."
      ],
      "metadata": {
        "id": "46F1fyY8IX3F"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Step 0. Downloading files**"
      ],
      "metadata": {
        "id": "f_51XAE3vAtg"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "First, download all genomes and proteins for complete bacterial genomes from GenBank. Outgroup databases could be larger, but for our purposes we're just going with all outgroup sequences in GenBank that are \"complete\".\n",
        "\n",
        "The outgroup databases (i.e., those consisting of non-E. coli genomes and proteins) are constructed with ATB and GenBank. ATB files were previously downloaded from their ftp site, but now it seems like they're hosted on a different platform.\n",
        "\n",
        "The E. coli genomes were available in the form of .gff files, which contained their sequence as well."
      ],
      "metadata": {
        "id": "ipNGOPRKIk3w"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jxCNc4nU02hp"
      },
      "outputs": [],
      "source": [
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes\n",
        "\n",
        "wget ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/assembly_summary.txt #Get GenBank assembly summary\n",
        "mv assembly_summary.txt 102824_GenBank_bacteria_assembly_summary.txt #rename the file with the date of download\n",
        "\n",
        "awk -F '\\t' '($12==\"Complete Genome\")' 102824_GenBank_bacteria_assembly_summary.txt | cut -f20 |\n",
        "awk -F '/' '{print $0\"/\"$NF\"_genomic.fna.gz\"}' > 102824_GenBank_bacteria_genome_URLs.txt #Extract URLs for all complete genomes\n",
        "awk -F '\\t' '($12==\"Complete Genome\")' 102824_GenBank_bacteria_assembly_summary.txt | cut -f20 |\n",
        "awk -F '/' '{print $0\"/\"$NF\"_protein.faa.gz\"}' > 102824_GenBank_bacteria_protein_URLs.txt #...and all corresponding proteins\n",
        "\n",
        "mkdir genomes && cp 102824_GenBank_bacteria_genome_URLs.txt genomes #prepare sub-directories to house these files\n",
        "cd genomes && wget -i 102824_GenBank_bacteria_genome_URLs.txt && gunzip *gz #download, unzip\n",
        "cd ..\n",
        "mkdir proteins && cp 102824_GenBank_bacteria_protein_URLs.txt proteins\n",
        "cd proteins && wget -i 102824_GenBank_bacteria_protein_URLs.txt && gunzip *gz #same with proteins"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Preparing outgroup databases**"
      ],
      "metadata": {
        "id": "_33Gc4PPv9ZE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "To construct outgroup databases, GenBank taxonomies cannot be relied upon. Each record must be validated to make sure there are no *E. coli* or *Escherichia* contaminants.\n",
        "\n",
        "(This needs not be done in case of ATB, in which species taxonomy designations were independently validated. I did it for them anyways, though)\n",
        "\n",
        "In the next code chunk, we generate ANI values between one focal strain of E. coli (K-12 MG1655) and all other bacteria.\n",
        "\n",
        "The focal strain was downloaded from here - https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000005845.2/\n",
        "\n",
        "And placed here - /stor/work/Ochman/hassan/protogene_extension/Ecoli_list/sequence_RS.fasta"
      ],
      "metadata": {
        "id": "4AWMGJIfYWPw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#For GenBank:\n",
        "\n",
        "mkdir /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_temp\n",
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_temp\n",
        "\n",
        "ls /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes | grep \"^GC\" | #get all genome files\n",
        "cut -f1,2 -d \"_\" | sed \"s/^/fastANI -q \\/stor\\/work\\/Ochman\\/hassan\\/protogene_extension\\/Ecoli_list\\/sequence_RS.fasta -r \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\//g\" | sed \"s/$/\\*fna/g\" | awk -F '/' '{print $0\" -o \"$NF}' | sed \"s/$/t/g\" | sed \"s/\\*fnat/_fastANI/g\" | #write fastANI code\n",
        "split -l 600 #split into 600-line chunks to run in parallel\n",
        "\n",
        "ls x* | sed \"s/^/bash /g\" > running.sh #all files resulting from the split command start with an \"x\"\n",
        "/stor/work/Ochman/hassan/tools/parallelize_run.sh #parallelize_run.sh is a helper script, described elsewhere\n",
        "\n",
        "#For ATB:\n",
        "#ATB files are stored in this directory: /stor/scratch/Ochman/hassan/0318_AllTheBacteria/\n",
        "#And all contigs from ATB are kept here: /stor/scratch/Ochman/hassan/0318_AllTheBacteria/AllTheBacteria_OG/\n",
        "#Only calculate fastANI values for records marked \"Escherichia\", to make sure none of them are very close to E. coli\n",
        "\n",
        "egrep \"Escherichia\" /stor/scratch/Ochman/hassan/0318_AllTheBacteria/hq_dataset.species_calls.tsv | grep -v \"Escherichia coli\" | cut -f1 | sed \"s/^/fastANI -q \\/stor\\/work\\/Ochman\\/hassan\\/protogene_extension\\/Ecoli_list\\/sequence_RS.fasta -r \\/stor\\/scratch\\/Ochman\\/hassan\\/0318_AllTheBacteria\\/AllTheBacteria_OG\\/\\*\\//g\" | sed \"s/$/\\.fa/g\" | awk -F '/' '{print $0\" -o \"$NF}' | sed \"s/$/t/g\" | sed \"s/\\.fat/_fastANI/g\" |\n",
        "split -l 100 - y #split into 100-line chunks to run in parallel. Make sure all resulting files start with a \"y\"\n",
        "\n",
        "ls y* | sed \"s/^/bash /g\" > running.sh #all resulting files start with an \"y\"\n",
        "/stor/work/Ochman/hassan/tools/parallelize_run.sh"
      ],
      "metadata": {
        "id": "jX8mZ3vuYTpO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Parse the fastANI results to identify only genuine outgroup genomes:"
      ],
      "metadata": {
        "id": "VvNcqyFXYPFr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ls | egrep \"fastANI$\" | sed \"s/^/cat /g\" | sed \"s/$/ >> Escherichia_all_fastANI/g\" | bash #Compile all fastANI files together\n",
        "#This includes both ATB and GenBank entries\n",
        "\n",
        "#Parse the fastANI results to make two-column files with accession ID and ANI\n",
        "#For GenBank:\n",
        "awk -F '\\t' '($2~\"GCA_\")' Escherichia_all_fastANI | cut -f2,3 |\n",
        "sed \"s/\\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\///g\" | sed 's/_/\\t/2' | cut -f1,3 > Escherichia_accession_ANI.tsv\n",
        "#For ATB:\n",
        "awk -F '\\t' '($2!~\"GCA_\")' Escherichia_all_fastANI | cut -f2,3 | rev | cut -f1 -d '/' | rev | sed \"s/\\.fa\\t/\\t/g\" >> Escherichia_accession_ANI.tsv\n",
        "sort -k1 Escherichia_accession_ANI.tsv -o Escherichia_accession_ANI.tsv #preparing a clean two-column tsv file - accession and ANI\n",
        "\n",
        "#There's a gap at 89% corresponding to Escherichia genus, exclude all accessions that's above this\n",
        "awk -F '\\t' '($2>89)' Escherichia_accession_ANI.tsv | grep \"GCA_\" | cut -f1 > Escherichia_genusexcluded_accesions.txt\n",
        "#we also exclude any record marked Escherichia in the assembly file, regardless of ANI:\n",
        "awk -F '\\t' '($12==\"Complete Genome\")' 102824_GenBank_bacteria_assembly_summary.txt | grep \"Escherichia\" |\n",
        "cut -f1 | sort -u >> Escherichia_genusexcluded_accesions.txt"
      ],
      "metadata": {
        "id": "yeaPTTvsYYMR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Make Escherichia-excluded outgroup genome and protein databases by excluding all Escherichia genomes from GB:"
      ],
      "metadata": {
        "id": "ZqLQRDKgjAqO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes\n",
        "cp Escherichia_temp/Escherichia_genusexcluded_accesions.txt .\n",
        "\n",
        "#Escherichia_excluded genomes:\n",
        "ls genomes/ | grep \"^GCA\" | grep -F -v -f Escherichia_genusexcluded_accesions.txt |\n",
        "sed \"s/^/cat \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\//g\" | sed \"s/$/ >> Escherichia_excluded.genomes.faa/g\" | bash\n",
        "#Escherichia_excluded proteins:\n",
        "ls proteins/ | grep \"^GCA\" | grep -F -v -f Escherichia_genusexcluded_accesions.txt |\n",
        "sed \"s/^/cat \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/proteins\\//g\" | sed \"s/$/ >> Escherichia_excluded.proteins.faa/g\" | bash\n",
        "#Escherichia_excluded ORFs:\n",
        "conda activate orfipy\n",
        "orfipy --start ATG,TTG,GTG --stop TGA,TAG,TAA --pep Escherichia_excluded.genomes.orfipy Escherichia_excluded.genomes.faa"
      ],
      "metadata": {
        "id": "cliruPA_jdwy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Similarly, in the next chunk make Ecoli-excluded outgroup genome and protein databases. This database includes all genomes in the Escherichia genus except those in E. coli.\n",
        "\n",
        "This database has two complications that the Escherichia-excluded one did not. First, this database includes genomes from both GB and ATB that need to be processed slightly differently.\n",
        "\n",
        "Second, the ATB genomes don't have annotations (they didn't when they were downloaded), so all E. coli-excluded Escherichia genomes are to be annotated with prodigal.\n",
        "\n",
        "More fundamentally, the strategy here differs from the previous code chunk: in the last one, we were excluding genomes. Here, we're excluding (<93) but also including (>89). The exclusion criteria is to remove E. coli genomes, while the inclusion criteria is to only retain genomes from other Escherichia species."
      ],
      "metadata": {
        "id": "j-NP2FNsjw0q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db\n",
        "\n",
        "#Get accessions of Ecoli-excluded genomes:\n",
        "#Genbank:\n",
        "awk -F '\\t' '($3<93&&$3>89)' Escherichia_accession_ANI.tsv | egrep -v \"Escherichia_coli|Escherichia_sp\" | grep \"^GCA\" | cut -f1 > Ecoli_excluded_accessions.GBRS.txt\n",
        "#ATB:\n",
        "awk -F '\\t' '($3<93&&$3>89)' Escherichia_accession_ANI.tsv | egrep -v \"Escherichia_coli|Escherichia_sp\" | grep -v \"^GCA\" | cut -f1 > Ecoli_excluded_accessions.ATB.txt\n",
        "\n",
        "#ls genomes/ | grep \"^GCA\" | grep -F -f Ecoli_excluded_accessions.GBRS.txt | sed \"s/^/cat \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\//g\" | sed \"s/$/ >> Ecoli_excluded.genomes.faa/g\" | bash\n",
        "#Concatenate Ecoli-excluded genomes into a multifasta database:\n",
        "#Genbank:\n",
        "sed \"s/^/cat \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\//g\" Ecoli_excluded_accessions.GBRS.txt | sed \"s/$/* >> Ecoli_excluded.genomes.faa/g\" | bash\n",
        "#ATB:\n",
        "sed \"s/^/cat \\/stor\\/scratch\\/Ochman\\/hassan\\/0318_AllTheBacteria\\/AllTheBacteria_OG\\/\\*\\//g\" Ecoli_excluded_accessions.ATB.txt | sed \"s/$/\\.fa >> Ecoli_excluded.genomes.faa/g\" | bash\n",
        "\n",
        "#Now for proteins. These genomes need to be annotated with prodigal.\n",
        "#We first make helper files\n",
        "sed \"s/^/ls \\/stor\\/scratch\\/Ochman\\/hassan\\/100724_Complete_Genomes\\/genomes\\//g\" Ecoli_excluded_accessions.GBRS.txt | sed \"s/$/*/g\" | bash > prodigal_interim\n",
        "sed \"s/^/ls \\/stor\\/scratch\\/Ochman\\/hassan\\/0318_AllTheBacteria\\/AllTheBacteria_OG\\/\\*\\//g\" Ecoli_excluded_accessions.ATB.txt | sed \"s/$/\\.fa/g\" | bash >> prodigal_interim\n",
        "\n",
        "#Run prodigal for all Ecoli-excluded genomes:\n",
        "sed \"s/^/prodigal -i /g\" prodigal_interim | sed \"s/$/ -f gff -o /g\" | awk -F '/' '{print $0$NF}' | sed \"s/ -f gff -o $//g\" | sed \"s/_genomic\\.fna$/\\.gff/g\" | sed \"s/\\.fa$/\\.gff/g\" | split -l100 #split them into 100 chunks for easier parallelization\n",
        "ls x* | sed \"s/^/bash /g\" > running.sh\n",
        "/stor/work/Ochman/hassan/tools/parallelize_run.sh\n",
        "\n",
        "#Convert the resulting gff files to gtf:\n",
        "for i in $(ls *gff | rev | cut -f2- -d \".\" | rev)\n",
        "do\n",
        "grep -v \"#\" \"$i\".gff | awk -F '\\t' '($3==\"CDS\")' | awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$4,$5,$6,$7,$8,$1}' | rev | cut -f 2- -d \".\" | rev | awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$4,$5,$6,$7,$8,\"transcript_id \\\"\"$9\"_\"NR\"\\\";gene_id \\\"\"$9\"_\"NR\"\\\";\"}'\n",
        "done\n",
        "\n",
        "#Pull sequences from genomes using the coordinates in gtf:\n",
        "#For GenBank:\n",
        "for i in $(ls gtf/GCA*gtf | rev | cut -f2- -d \".\" | rev | cut -f2- -d \"/\")\n",
        "do\n",
        "cat gtf/\"$i\".gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes/\"$i\"_genomic.fna -bed - > cds/\"$i\".cds.faa\n",
        "done\n",
        "\n",
        "#For ATB:\n",
        "for i in $(ls gtf/SA*gtf | rev | cut -f2- -d \".\" | rev | cut -f2- -d \"/\")\n",
        "do\n",
        "cat gtf/\"$i\".gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/scratch/Ochman/hassan/0318_AllTheBacteria/AllTheBacteria_OG/*/$i.fa -bed - > cds/$i.cds.faa\n",
        "done\n",
        "\n",
        "#cds to protein:\n",
        "for i in $(ls *.cds.faa | rev | cut -f3- -d '.' | rev | cut -f2- -d \"/\")\n",
        "do\n",
        "/stor/work/Ochman/hassan/tools/faTrans -stop $i.cds.faa $i.protein.faa\n",
        "done\n",
        "\n",
        "#Compile these together to make Ecoli_excluded proteins:\n",
        "cat *protein.faa >> Ecoli_excluded.proteins.faa\n",
        "\n",
        "#Ecoli_excluded ORFs:\n",
        "orfipy --start ATG,TTG,GTG --stop TGA,TAG,TAA --pep Ecoli_excluded.genomes.orfipy Ecoli_excluded.genomes.faa"
      ],
      "metadata": {
        "id": "MeBMtorbkBls"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Preparing pangenome databases**\n",
        "\n"
      ],
      "metadata": {
        "id": "gnb1boQkCmbA"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Steps:\n",
        "\n",
        "1.   Extract the genomes from the .gff files. The gffs were available already\n",
        "2.   Annotate the genomes with genemarks2 (gffs are already the product of prodigal)\n",
        "3.   Convert both types of .gff files to .gtf files\n",
        "4. Extract cds using gtf\n",
        "5. Convert cds to proteins\n",
        "6. Only retain those genomes that are in the 44 lineages under study\n",
        "7. Make ORF-based database"
      ],
      "metadata": {
        "id": "MR54246Rvoly"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#move the gffs to appropriate directory and rename them to indicate annotation source:\n",
        "\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gffs/\n",
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gffs/\n",
        "ls *gff | rev | cut -f2- -d '.' | rev | awk '{print \"mv \"$0\".gff \"$0\".prodigal.gff\"}' | bash\n",
        "\n",
        "#Step-1\n",
        "\n",
        "#Extract genomes from gffs:\n",
        "for i in $(ls *gff | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "awk '/##FASTA/ {found=1; next} found' \"$i\".prodigal.gff > \"$i\".fasta\n",
        "done\n",
        "\n",
        "#Move the genome files to appropriate directories:\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes\n",
        "mv *fasta /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\n",
        "\n",
        "#Step-2\n",
        "\n",
        "#Annotate the genomes with genemarks2:\n",
        "#genemarks2 requires the \"key\" file in the directory in which it's run, plus in the home directory\n",
        "\n",
        "for i in $(ls /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/*fasta | rev | cut -f1 -d '/' | cut -f2- -d '.' | rev)\n",
        "do\n",
        "/stor/work/Ochman/hassan/tools/gms2_linux_64/gms2.pl --seq /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\"$i\".fasta --genome-type bacteria --gcode 11 --output \"$i\".genemarks2.gff --format gff3\n",
        "done\n",
        "\n",
        "#Step-3\n",
        "\n",
        "#Convert both gff types into gtfs:\n",
        "\n",
        "for i in $(ls *prodigal.gff | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "awk '/##FASTA/{exit} {print}' \"$i\".prodigal.gff |\n",
        "grep -v \"#\" | awk -F '\\t' '($3==\"CDS\")' |\n",
        "cut -f1 -d ';' | sed \"s/ID=//g\" |\n",
        "awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$4,$5,$6,$7,$8,\"transcript_id \\\"\"$9\"\\\";\\gene_id \\\"\"$9\"\\\";\"}' > \"$i\".prodigal.gtf\n",
        "done\n",
        "\n",
        "for i in $(ls *genemarks2.gff | rev | cut -f3- -d \".\" | rev)\n",
        "do\n",
        "awk -F '\\t' '($3==\"CDS\")' \"$i\".genemarks2.gff |\n",
        "cut -f1 -d \";\" | sed \"s/ID=/genemarks2_/g\" |\n",
        "awk -F '\\t' '{OFS=FS}{print $1,$2,$3,$4,$5,$6,$7,$8,\"transcript_id \\\"\"$9\"\\\";\\gene_id \\\"\"$9\"\\\";\"}' > \"$i\".genemarks2.gtf\n",
        "done\n",
        "\n",
        "#Move them to appropriate directories:\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/prodigal_gtfs\n",
        "mkdir /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/genemarks2_gtfs\n",
        "mv /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/*.prodigal.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/prodigal_gtfs\n",
        "mv /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/*.genemarks2.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/genemarks2_gtfs\n",
        "\n",
        "#Step-4\n",
        "\n",
        "#Now to extract the corresponding sequences and put concatenate them to make genome and protein files\n",
        "\n",
        "mkdir pangenome_database_interim\n",
        "\n",
        "#gtf to cds:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/ | grep -v \"filter\" | rev | cut -f3- -d \".\" | rev) #the \"grep -v filter\" is needed to account for other files that were created in other steps of the process\n",
        "do\n",
        "cat ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\"$i\".fasta -bed - > pangenome_database_interim/\"$i\".prodigal.CDS.faa\n",
        "done\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/genemarks2_gtfs/ | grep -v \"filter\" | rev | cut -f3- -d \".\" | rev) #the \"grep -v filter\" is needed to account for other files that were created in other steps of the process\n",
        "do\n",
        "cat ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\"$i\".fasta -bed - > pangenome_database_interim/\"$i\".genemarks2.CDS.faa\n",
        "seqkit fx2tab pangenome_database_interim/\"$i\".genemarks2.CDS.faa | sed \"s/\\t/,/g\" | egrep \",ATG|,TTG|,GTG\" | egrep \"TAA,|TGA,|TAG,\" | sed \"s/^/>/g\" | sed \"s/,$//g\" | sed \"s/,/\\n/g\" > interim && mv interim pangenome_database_interim/\"$i\".genemarks2.CDS.faa\n",
        "done\n",
        "\n",
        "#Step-5\n",
        "\n",
        "#cds to proteins:\n",
        "\n",
        "for i in $(ls pangenome_database_interim/*faa | rev | cut -f3- -d '.' | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "/stor/work/Ochman/hassan/tools/faTrans -stop pangenome_database_interim/\"$i\".CDS.faa pangenome_database_interim/\"$i\".prot.faa\n",
        "done\n",
        "\n",
        "#Step-6\n",
        "\n",
        "#Only retain proteins corresponding to the 44 lineages under study:\n",
        "#the /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv file was extracted from supplement\n",
        "\n",
        "for i in $(awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | cut -f1)\n",
        "do\n",
        "sed \"s/>/>\"$i\"@/g\" pangenome_database_interim/\"$i\"*prot.faa >> Ecoli.all.prot.pangenomedb.faa #Each protein is marked with the genome ID, so we know their source\n",
        "done\n",
        "\n",
        "#Step-7\n",
        "\n",
        "#To make ORF-based database, first conncatenate all genomes in the 44 lineages under study:\n",
        "for i in $(awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | cut -f1)\n",
        "do\n",
        "sed \"s/>/>\"$i\"@/g\" Ecoli_genomes/\"$i\".fasta >> Ecoli.all.genomes.faa #Each contig is marked with the genome ID, so we know their source\n",
        "done\n",
        "\n",
        "#Extract proteins encoded from all ORFs:\n",
        "orfipy --start ATG,TTG,GTG --stop TGA,TAG,TAA --pep Ecoli.all.genomes.orfipy Ecoli.all.genomes.faa\n"
      ],
      "metadata": {
        "id": "YilpydhOsuuO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we convert each of these genomes, proteins, and translated ORF files into blast or diamond databases."
      ],
      "metadata": {
        "id": "ACkfCYgiuDk8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Escherichia and Ecoli_excluded genome blast databases:\n",
        "makeblastdb -in Escherichia_excluded.genomes.faa -dbtype nucl -out Escherichia_excluded.genomes\n",
        "makeblastdb -in Ecoli_excluded.genomes.faa -dbtype nucl -out Ecoli_excluded.genomes\n",
        "makeblastdb -in Ecoli.all.genomes.faa -dbtype nucl -out Ecoli.all.genomes\n",
        "#Escherichia, Ecoli_excluded and pangenome protein diamond databses:\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in Escherichia_excluded.proteins.faa --db Escherichia_excluded.proteins\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in Ecoli_excluded.proteins.faa --db Ecoli_excluded.proteins\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in Ecoli.all.prot.pangenomedb.faa --db Ecoli.all.prot.pangenomedb\n",
        "#Escherichia, Ecoli_excluded and pangenome translated ORF diamond databases:\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in orfipy_Escherichia_excluded.genomes.faa_out/Escherichia_excluded.genomes.orfipy --db Escherichia_excluded.genomes.orfipy.prot\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in orfipy_Ecoli_excluded.genomes.faa_out/Ecoli_excluded.genomes.orfipy --db Ecoli_excluded.genomes.orfipy.prot\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in orfipy_Ecoli.all.genomes.faa_out/Ecoli.all.genomes.orfipy --db Ecoli.all.orfipy\n"
      ],
      "metadata": {
        "id": "Vu7KaMbpuJ6e"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "There's one last convenience step to take care of. In a subsequent step, we run samtools to get sequences based on coordinates, and that step is significantly sped up if we place each contig in individual files named after those contigs.\n",
        "\n",
        "For this, we prepare the individual_genomes directory."
      ],
      "metadata": {
        "id": "9RtjKxGxiInf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Make helper directories for each of our three databases\n",
        "\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/\n",
        "mkdir individual_genomes\n",
        "cd individual_genomes\n",
        "mkdir Escherichia_excluded\n",
        "mkdir Ecoli_excluded\n",
        "mkdir Ecoli_all\n",
        "\n",
        "#Move all genome files here\n",
        "cp Escherichia_excluded.genomes.faa individual_genomes/Escherichia_excluded\n",
        "cp Ecoli_excluded.genomes.faa individual_genomes/Ecoli_excluded\n",
        "cp Ecoli.all.genomes.faa individual_genomes/Ecoli_all\n",
        "\n",
        "\n",
        "cd Escherichia_excluded\n",
        "#Linearize files first\n",
        "seqkit fx2tab Escherichia_excluded.genomes.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" > temp && mv temp Escherichia_excluded.genomes.faa\n",
        "#Split the file into two-line files, first line identifier, second line sequence\n",
        "split -l 2 Escherichia_excluded.genomes.faa\n",
        "#This results in a bunch of files start with \"x\"\n",
        "#Rename them to their sequence name\n",
        "for i in x*\n",
        "do\n",
        "rename=$(grep \"^>\" $i | cut -f1 -d \" \" | tr -d \">\")\n",
        "mv $i $rename\n",
        "done\n",
        "cd ..\n",
        "\n",
        "#Same with the Ecoli_excluded database\n",
        "\n",
        "cd Ecoli_excluded\n",
        "seqkit fx2tab Ecoli_excluded.genomes.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" > temp && mv temp Ecoli_excluded.genomes.faa\n",
        "split -l 2 Ecoli_excluded.genomes.faa\n",
        "for i in x*\n",
        "do\n",
        "rename=$(grep \"^>\" $i | cut -f1 -d \" \" | tr -d \">\")\n",
        "mv $i $rename\n",
        "done\n",
        "cd ..\n",
        "\n",
        "#Same with the Ecoli_all database\n",
        "\n",
        "cd Ecoli_all\n",
        "seqkit fx2tab Ecoli.all.genomes.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" > temp && mv temp Ecoli.all.genomes.faa\n",
        "split -l 2 Ecoli.all.genomes.faa\n",
        "for i in x*\n",
        "do\n",
        "rename=$(grep \"^>\" $i | cut -f1 -d \" \" | tr -d \">\")\n",
        "mv $i $rename\n",
        "done\n",
        "cd ..\n",
        "\n",
        "#Move them all back to main individual_genomes directory\n",
        "mv Escherichia_excluded/* .\n",
        "mv Ecoli_excluded/* .\n",
        "mv Ecoli_all/* ."
      ],
      "metadata": {
        "id": "-emEuRUzf95Y"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Download and prepare the nr database**"
      ],
      "metadata": {
        "id": "eRM0u5i1EYPx"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Finally, for a sanity check step near the end of the pipeline, I also search the proteins against the nonredundant protein database. In the chunk below I download and prepare the database by incorporating the taxonomy information."
      ],
      "metadata": {
        "id": "67dNEKk2EfaN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "mkdir /stor/scratch/Ochman/hassan/nr\n",
        "cd /stor/scratch/Ochman/hassan/nr\n",
        "\n",
        "#Download the nr file\n",
        "\n",
        "wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz\n",
        "\n",
        "#Download associated taxonomy information files\n",
        "\n",
        "wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz\n",
        "gunzip prot.accession2taxid.FULL.gz\n",
        "\n",
        "wget https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz\n",
        "tar -zxvf taxdump.tar.gz\n",
        "\n",
        "#Make diamond database:\n",
        "diamond makedb \\\n",
        "    --in nr.gz \\\n",
        "    --db 03272024_nr \\\n",
        "    --taxonmap prot.accession2taxid.FULL \\\n",
        "    --taxonnames names.dmp \\\n",
        "    --taxonnodes nodes.dmp"
      ],
      "metadata": {
        "id": "mdIWirlpEuUV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Extract taxonomy information for each database entry**"
      ],
      "metadata": {
        "id": "Jqz3kvdgv5wn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "For subsequent steps, it's important to get the taxonomic information to each target genome and protein. The goal is to make three-column tsv files, with the first column listing genome accession, the second column listing contig or protein ID, and the third column listing appropriate taxonomic information -- genus in case of non-Escherichia records, species in case of non-E. coli records, and lineage in case of pangenome records."
      ],
      "metadata": {
        "id": "QcRyuOK3vMTe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/\n",
        "\n",
        "#First, make a two-column file with accession number and species name:\n",
        "awk -F '\\t' '($12==\"Complete Genome\")' 102824_GenBank_bacteria_assembly_summary.txt | cut -f1,8 | sed \"s/ /_/g\" > accession_taxonomy.tsv\n",
        "sort -k1 accession_taxonomy.tsv -o accession_taxonomy.tsv\n",
        "\n",
        "#For proteins and accessions:\n",
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/proteins\n",
        "#Two-column file with accession and protein ID:\n",
        "for i in $(ls | grep \"GCA\" | cut -f 1,2 -d \"_\"); do grep \"^>\" \"$i\"*protein.faa | cut -f1 -d \" \" | tr -d \">\" | sed \"s/^/\"$i\"\\t/g\" >> accession_proteinID.tsv; done\n",
        "#Join this with the file containing accession and taxonomy:\n",
        "sort -k1 accession_proteinID.tsv | join -1 1 -2 1 - ../accession_taxonomy.tsv | sed \"s/ /\\t/g\" > accession_proteinID_taxonomy.tsv #tie this to taxonomy\n",
        "sort -k2 /stor/scratch/Ochman/hassan/100724_Complete_Genomes/proteins/accession_proteinID_taxonomy.tsv -o /stor/scratch/Ochman/hassan/100724_Complete_Genomes/proteins/accession_proteinID_taxonomy.tsv\n",
        "\n",
        "#For genomes and accessions:\n",
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes\n",
        "#Two-column file with accession and genome ID:\n",
        "for i in $(ls | grep \"GCA\" | cut -f 1,2 -d \"_\"); do grep \"^>\" \"$i\"*genomic.fna | cut -f1 -d \" \" | tr -d \">\" | sed \"s/^/\"$i\"\\t/g\" >> accession_genomeID.tsv; done\n",
        "#Join this with the file containing accession and taxonomy:\n",
        "sort -k1 accession_genomeID.tsv | join -1 1 -2 1 - ../accession_taxonomy.tsv | sed \"s/ /\\t/g\" > accession_genomeID_taxonomy.tsv #tie this to taxonomy\n",
        "sort -k2 /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes/accession_genomeID_taxonomy.tsv -o /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes/accession_genomeID_taxonomy.tsv\n",
        "\n",
        "#Only extract those rows corresponding to genome IDs in the GenBank Escherichia_excluded database:\n",
        "cd /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db\n",
        "grep \"^>\" Escherichia_excluded.genomes.faa | cut -f1 -d \" \" | tr -d \">\" > Escherichia_excluded_accessions.GBRS.txt\n",
        "cat Escherichia_excluded_accessions.GBRS.txt | grep -w -F -f - /stor/scratch/Ochman/hassan/100724_Complete_Genomes/genomes/accession_genomeID_taxonomy.tsv | sort -k2 > Ecoli_extragenus_genome_contig_taxa.tsv\n",
        "#Only retain genus names (and get rid of non-useful prefixes) — that's what makes it \"reduced\":\n",
        "sed \"s/^\\[//g\" Ecoli_extragenus_genome_contig_taxa.tsv | sed \"s/uncultured_//g\" | sed \"s/Candidatus_//g\" | sed \"s/^'//g\" | sed \"s/candidatus_//g\" | awk '$3 ~ /^[A-Z]/' | awk '{sub(/_.*/, \"\", $3)}1' | sed \"s/ /\\t/g\" | sort -k2 > Ecoli_extragenus_genome_contig_taxa.reduced.tsv\n",
        "#Apparently some E_coli IDs snuck into the database...\n",
        "grep -v \"Escherichia\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.tsv > /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.noescherichia.tsv\n",
        "\n",
        "#Only extract those rows corresponding to genome IDs in the GenBank Escherichia_excluded database:\n",
        "grep \"^>\" Escherichia_excluded.proteins.faa | cut -f1 -d \" \" | tr -d \">\" > Escherichia_db/Escherichia_excluded_accessions.proteins.GBRS.txt\n",
        "cat Escherichia_excluded_accessions.proteins.GBRS.txt | grep -w -F -f - proteins/accession_proteinID_taxonomy.tsv | sort -k2 > Ecoli_extragenus_protein_contig_taxa.tsv\n",
        "#Only retain genus names (and get rid of non-useful prefixes) — that's what makes it \"reduced\":\n",
        "sed \"s/^\\[//g\" Ecoli_extragenus_protein_contig_taxa.tsv | sed \"s/uncultured_//g\" | sed \"s/Candidatus_//g\" | sed \"s/^'//g\" | sed \"s/candidatus_//g\" | awk '$3 ~ /^[A-Z]/' | awk '{sub(/_.*/, \"\", $3)}1' | sed \"s/ /\\t/g\" | sort -k2 > Ecoli_extragenus_protein_contig_taxa.reduced.tsv\n",
        "#Apparently some E_coli IDs snuck into the database...\n",
        "grep -v \"Escherichia\" Ecoli_extragenus_protein_contig_taxa.reduced.tsv > /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_protein_contig_taxa.reduced.noescherichia.tsv\n",
        "\n",
        "#These Escherichia contaminants need to be removed from the blast search results\n",
        "#We store their accession in a separate file for this process:\n",
        "grep \"Escherichia\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_protein_contig_taxa.reduced.tsv | cut -f2 > /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/all_Escherichia_slippedby_IDs.txt\n",
        "grep \"Escherichia\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.tsv | cut -f2 >> /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering/all_Escherichia_slippedby_IDs.txt\n",
        "\n",
        "#Same with intragenus (i.e., E. coli-excluded), and for genomes:\n",
        "#For GenBank:\n",
        "cat Ecoli_excluded_accessions.GBRS.txt | grep -w -F -f - genomes/accession_genomeID_taxonomy.tsv | sort -k2 > Ecoli_intragenus_genome_contig_taxa.tsv\n",
        "#For ATB:\n",
        "grep \"^>\" Ecoli_excluded.genomes.faa | awk '($1~\"contig\")' | cut -f1 -d \" \" | sort -u | tr -d \">\" | awk -F '.' '{print $1,$0}' | sort -nrk1 > temp\n",
        "cut -f1 -d \" \" temp | grep -w -F -f - ../0318_AllTheBacteria/hq_dataset.species_calls.tsv | sed \"s/ /_/g\" | sort -nrk1 | join -1 1 -2 1 - temp | awk '{print $1\"\\t\"$3\"\\t\"$2}' >> Ecoli_intragenus_genome_contig_taxa.tsv\n",
        "\n",
        "#Same with intragenus (i.e., E. coli-excluded), and for protein:\n",
        "#GenBank:\n",
        "#ProteinID-Accession tsv:\n",
        "grep \"^>\" Escherichia_db/proteins/*protein.faa | grep \"GCA\" | rev | cut -f1 -d \"/\" | rev | sed \"s/_/\\t/2\" | sed \"s/>/\\t/g\" | cut -f1,3 | rev | cut -f2- -d \"(\" | rev > Ecoli_intragenus_genome_protein.tsv\n",
        "#ATB:\n",
        "grep \"^>\" Escherichia_db/proteins/*protein.faa | grep -v \"GCA\" | rev | cut -f1 -d \"/\" | rev | sed \"s/\\./\\t/1\" | sed \"s/>/\\t/g\" | cut -f1,3 | rev | cut -f2- -d \"(\" | rev >> Ecoli_intragenus_genome_protein.tsv\n",
        "sort -k1 Ecoli_intragenus_genome_protein.tsv -o Ecoli_intragenus_genome_protein.tsv\n",
        "#Join this with the file containing genome accession and species:\n",
        "sort -k1 Ecoli_intragenus_genome_contig_taxa.tsv | join -1 1 -2 1 - Ecoli_intragenus_genome_protein.tsv | awk '{print $1\"\\t\"$4\"\\t\"$3}' > Ecoli_intragenus_genome_protein_taxa.tsv\n",
        "\n",
        "#Slight fixes to some species names:\n",
        "sed -i \"s/_KF1//g\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_contig_taxa.tsv\n",
        "sed -i \"s/_KF1//g\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_protein_taxa.tsv\n",
        "sed -i \"s/_ATCC_35469//g\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_contig_taxa.tsv\n",
        "sed -i \"s/_ATCC_35469//g\" /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_protein_taxa.tsv\n",
        "\n",
        "#For pangenome:\n",
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3\n",
        "\n",
        "#genomes:\n",
        "grep \"^>\" Ecoli.all.genomes.faa | tr -d \">\" | awk -F '@' '{print $1\"\\t\"$0}' | sort -k1 > temp\n",
        "awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | awk -F '\\t' '{print $1\"\\tEcoli@\"$5}' | sort -k1 | join -1 1 -2 1 - temp | awk '{print $1\"\\t\"$3\"\\t\"$2}' > Ecoli_pangenome_genome_contig_taxa.tsv\n",
        "\n",
        "#proteins:\n",
        "grep \"^>\" Ecoli.all.prot.faa | tr -d \">\" | awk -F \"@\" '{print $1\"\\t\"$0}' | sort -k1 > temp\n",
        "awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | awk -F '\\t' '{print $1\"\\tEcoli@\"$5}' | sort -k1 | join -1 1 -2 1 - temp | awk '{print $1\"\\t\"$3\"\\t\"$2}' > Ecoli_pangenome_genome_protein_taxa.tsv\n",
        "\n",
        "cp Ecoli_pangenome_genome_contig_taxa.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes\n",
        "cp Ecoli_pangenome_genome_protein_taxa.tsv /stor/scratch/Ochman/hassan/100724_Complete_Genomes\n",
        "sort -k2 /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_protein_taxa.tsv -o /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_protein_taxa.tsv\n",
        "sort -k2 /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv -o /stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv"
      ],
      "metadata": {
        "id": "ExWoV6-hvbo0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Final set of databases:\n",
        "\n",
        "#Genomes:\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.genomes\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Ecoli_excluded.genomes\n",
        "/stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli.all.genomes\n",
        "\n",
        "#Proteins:\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.proteins.dmnd\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Ecoli_excluded.proteins.dmnd\n",
        "/stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli.all.prot.dmnd\n",
        "\n",
        "#ORFs:\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Escherichia_excluded.proteins.dmnd\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Escherichia_db/Ecoli_excluded.proteins.dmnd\n",
        "/stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli.all.orfipy.dmnd\n",
        "\n",
        "#nr:\n",
        "/stor/scratch/Ochman/hassan/nr.dmnd\n",
        "\n",
        "#Final set of TSV files that contain genome+protein IDs and corresponding taxonomy:\n",
        "\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_genome_contig_taxa.reduced.noescherichia.tsv\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_contig_taxa.tsv\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_contig_taxa.tsv\n",
        "\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_extragenus_protein_contig_taxa.reduced.noescherichia.tsv\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_intragenus_genome_protein_taxa.tsv\n",
        "/stor/scratch/Ochman/hassan/100724_Complete_Genomes/Ecoli_pangenome_genome_protein_taxa.tsv"
      ],
      "metadata": {
        "id": "tCQuh6WOgFgr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "oPduIuQTxmoG"
      }
    }
  ]
}