{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "E5HeSHbkTTyx",
        "jWbBEawiZ06A"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "##**Filtering proteins**\n",
        "\n",
        "Our point of departure are the prodigal and genemarks2 gtf files generated in the first step of this pipeline (see preparing_databases.sh). These files are kept in /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/\n",
        "\n",
        "We exclude genes that are:\n",
        "\n",
        "1. Only found at contig edges -- i.e., have no genes upstream or downstream; or\n",
        "2. Only annotated by one of the two annotation programs; or\n",
        "3. Lacking conventional starts and/or stops."
      ],
      "metadata": {
        "id": "dM1OqAJ8-pQh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Preparing query proteins**\n",
        "\n",
        "To detect ORFan genes in a bacterial species, the first step is to screen all proteins encoded by that species. In this project, we're only interested in well-established proteins that are picked out by two annotation programs -- prodigal and genemarks2, in our case. Furthermore, we'll be excluding proteins that are exclusively located at the edge of contigs, or genes that don't have an ATG/TTG/GTG start or an TAG/TGA/TAA stop.\n",
        "\n",
        "After collecting these proteins from a set of genomes, we have to cluster them based on similarity so we're ultimately left with a manageable number of gene families. It's not practical to search millions of proteins across the database. Instead, we cluster them and pick just one representative sequence from the cluster.\n",
        "\n",
        "This notebook is divided into two halves: the filtering step, and the clustering step."
      ],
      "metadata": {
        "id": "hqlbDg7920Qz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Filter 1. Genes at edges**"
      ],
      "metadata": {
        "id": "cVfo8deiFGWQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering\n",
        "\n",
        "#For prodigal, pick out the edge genes:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*gtf | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "awk -F'\\t' 'BEGIN{OFS=\"\\t\"}{if(!($1 in seen)){order[++n]=$1;seen[$1]=1;first[$1]=$0}last[$1]=$0}END{for(i=1;i<=n;i++){k=order[i];print first[k];if(first[k]!=last[k])print last[k]}}' ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf | cut -f2 -d \"\\\"\" | sed \"s/^/\"$i\",/g\" >> prodigal_edge_genes.csv\n",
        "done\n",
        "\n",
        "#For genemarks2, pick out the edge genes:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/genemarks2_gtfs/*gtf | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "awk -F'\\t' 'BEGIN{OFS=\"\\t\"}{if(!($1 in seen)){order[++n]=$1;seen[$1]=1;first[$1]=$0}last[$1]=$0}END{for(i=1;i<=n;i++){k=order[i];print first[k];if(first[k]!=last[k])print last[k]}}' ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf | cut -f2 -d \"\\\"\" | sed \"s/^/\"$i\",/g\" >> genemarks2_edge_genes.csv\n",
        "done\n",
        "\n",
        "#The files are two-column csvs, with first column = genome name, second column = gene/protein name"
      ],
      "metadata": {
        "id": "CTwqbGx5-n6e"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Filter 2. Lacking dual annotation**"
      ],
      "metadata": {
        "id": "HBlte22qdWQ_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#For both cases, identify genes that aren't picked out by both annotation programs\n",
        "#To this end, identify what genes are annotated by what program:\n",
        "\n",
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gffs/\n",
        "\n",
        "for i in $(ls *genemarks2.gff | rev | cut -f3- -d \".\" | rev) #Loop using file basenames\n",
        "do\n",
        "#The first line makes non-redundant gtf files by first merging prodigal and genemarks2 annotations\n",
        "#Then checking to see if any gene has the same contig and stop codon information\n",
        "#In which case, the longer one is retained\n",
        "cat /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf | awk '{stop = ($7 == \"+\") ? $5 : $4; feature_length = $5 - $4; key = $1 \":\" stop; if (!(key in data) || feature_length > data[key]) {data[key] = feature_length; lines[key] = $0}} END {for (k in lines) {print lines[k]}}' > \"$i\".nr.gtf\n",
        "#The next four lines take the non-redundant gtf file as input\n",
        "#Checks to see if any entry in the prodigal/genemarks2 gtf shares its contig and stop codon information\n",
        "#Based on this, it evaluates if the gene is found by prodigal and/or genemarks2\n",
        "awk -F'\\t' 'NR==FNR { seen[$1, $5, $7] = $9; next } ($7 == \"+\" && ($1, $5, $7) in seen) { print seen[$1, $5, $7], \"prodigal\" }' \"$i\".nr.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf > \"$i\".tmp\n",
        "awk -F'\\t' 'NR==FNR { seen[$1, $4, $7] = $9; next } ($7 == \"-\" && ($1, $4, $7) in seen) { print seen[$1, $4, $7], \"prodigal\" }' \"$i\".nr.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf >> \"$i\".tmp\n",
        "awk -F'\\t' 'NR==FNR { seen[$1, $5, $7] = $9; next } ($7 == \"+\" && ($1, $5, $7) in seen) { print seen[$1, $5, $7], \"genemarks2\" }' \"$i\".nr.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf >> \"$i\".tmp\n",
        "awk -F'\\t' 'NR==FNR { seen[$1, $4, $7] = $9; next } ($7 == \"-\" && ($1, $4, $7) in seen) { print seen[$1, $4, $7], \"genemarks2\" }' \"$i\".nr.gtf /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf >> \"$i\".tmp\n",
        "rev \"$i\".tmp | cut -f-2 -d \" \" | rev | tr -d \"\\\"\" | sed \"s/;//g\" | awk '{print $2\"\\t\"$1}' | awk -F'\\t' '{ values[$2] = (values[$2] == \"\" ? $1 : values[$2] \", \" $1) } END { for (value in values) { print value \"\\t\" values[value] } }' | sed \"s/ //g\" > \"$i\".annotcomp.tsv\n",
        "done\n",
        "\n",
        "#Concatenate the annotation presence/absence info for each gene into a singular file\n",
        "#Don't forget to mark protein names with their genome info so we know where they came from\n",
        "for i in *annotcomp.tsv; do cat $i | sed \"s/^/\"$i\"@/g\"; done | sed \"s/.annotcomp.tsv//g\" >> all_annotcomp.tsv\n",
        "\n",
        "#Pick out ones without dual annotation:\n",
        "awk -F '\\t' '($2!~\",\")' ../Ecoli_gffs/all_annotcomp.tsv | cut -f1 | sort -u | sed \"s/@/,/g\" > genes_with_oneannotation.csv"
      ],
      "metadata": {
        "id": "CCUzS2EddVVb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **3. Lacking normal start/stop codons**"
      ],
      "metadata": {
        "id": "QlJJ1wwEdiNL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering\n",
        "\n",
        "#Genes that begin with other than ATG, TTG or GTG:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*gtf | grep -v \"filter\" | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf | gtf2bed | bedtools getfasta -s -name -fi ../Ecoli_genomes/\"$i\".fasta -bed - |\n",
        "seqkit fx2tab | sed \"s/\\t/,/g\" | egrep -v \",ATG|,TTG|,GTG\" | cut -f1 -d \"(\" | sed \"s/^/\"$i\",/g\"\n",
        "done >> prodigal_genes_without_start.csv\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/genemarks2_gtfs/*gtf | grep -v \"filter\" | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf | gtf2bed | bedtools getfasta -s -name -fi ../Ecoli_genomes/\"$i\".fasta -bed - |\n",
        "seqkit fx2tab | sed \"s/\\t/,/g\" | egrep -v \",ATG|,TTG|,GTG\" | cut -f1 -d \"(\" | sed \"s/^/\"$i\",/g\"\n",
        "done >> genemarks2_genes_without_start.csv\n",
        "\n",
        "#Those missing stops:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*gtf | grep -v \"filter\" | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf | gtf2bed | bedtools getfasta -s -name -fi ../Ecoli_genomes/\"$i\".fasta -bed - |\n",
        "seqkit fx2tab | sed \"s/\\t/,/g\" | egrep -v \"TGA,|TAA,|TAG,\" | cut -f1 -d \"(\" | sed \"s/^/\"$i\",/g\"\n",
        "done >> prodigal_genes_without_stop.csv\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/genemarks2_gtfs/*gtf | grep -v \"filter\" | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf | gtf2bed | bedtools getfasta -s -name -fi ../Ecoli_genomes/\"$i\".fasta -bed - |\n",
        "seqkit fx2tab | sed \"s/\\t/,/g\" | egrep -v \"TGA,|TAA,|TAG,\" | cut -f1 -d \"(\" | sed \"s/^/\"$i\",/g\"\n",
        "done >> genemarks2_genes_without_stop.csv"
      ],
      "metadata": {
        "id": "qegO6rYvdpHI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **4. Remove unwanted genes**\n",
        "\n",
        "Now filter each .gtf file by removing edge, single-annotation, unconventional-startstop genes"
      ],
      "metadata": {
        "id": "jrU77rD_iXej"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cd /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/rethinking_clustering\n",
        "\n",
        "#First exclude genes from prodigal gtfs:\n",
        "\n",
        "cat genes_with_oneannotation.csv | grep -v \"genemarks2\" | #Only select prodigal-annotated genes without dual annotation\n",
        "cat - prodigal_genes_without_start.csv prodigal_genes_without_stop.csv prodigal_edge_genes.csv | #Add to this the non-ORF genes, the edge genes\n",
        "sort -u > all_prodigal_exclude.csv #Complete exclusion list\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*gtf | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "gene_name=$(awk -v var=\"$i\" -F ',' '($1==var)' all_prodigal_exclude.csv | cut -f2 -d ',') #Gene name in second column\n",
        "grep -v -w -F \"$gene_name\" ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.gtf >> ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.filtered.gtf #Exclude the genes from the corresponding .gtf\n",
        "done\n",
        "\n",
        "#Likewise with genemarks2:\n",
        "\n",
        "cat genes_with_oneannotation.csv | grep \"genemarks2\" |\n",
        "cat - genemarks2_genes_without_start.csv genemarks2_genes_without_stop.csv genemarks2_edge_genes.csv |\n",
        "sort -u > all_genemarks2_exclude.csv\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/genemarks2_gtfs/*gtf | rev | cut -f3- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "gene_name=$(awk -v var=\"$i\" -F ',' '($1==var)' all_genemarks2_exclude.csv | cut -f2 -d ',')\n",
        "grep -v -w -F \"$gene_name\" ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.gtf >> ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.filtered.gtf\n",
        "done"
      ],
      "metadata": {
        "id": "M3ua5WigicaL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "For each genome, we combine their filtered annotations to yield nonredundant .gtf files:"
      ],
      "metadata": {
        "id": "7rOvK3BPmFfI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*prodigal.filtered.gtf | rev | cut -f4- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/prodigal_gtfs/\"$i\".prodigal.filtered.gtf ../Ecoli_gtfs/genemarks2_gtfs/\"$i\".genemarks2.filtered.gtf | #concatenating both .gtf files\n",
        "#As before, we concatenate the two filtered .gtf files and remove redundancy by retaining the longer gene corresponding to same stop codon in the same contig\n",
        "awk '{stop = ($7 == \"+\") ? $5 : $4; feature_length = $5 - $4; key = $1 \":\" stop; if (!(key in data) || feature_length > data[key]) {data[key] = feature_length; lines[key] = $0}} END {for (k in lines) {print lines[k]}}' > ../Ecoli_gtfs/\"$i\".filtered.nr.gtf\n",
        "done\n"
      ],
      "metadata": {
        "id": "rNVekJcQmQid"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **5. Make corresponding CDS and protein files from 44 lineages**\n",
        "\n",
        "Finally, it's time to:\n",
        "\n",
        "1.   Extract CDS corresponding to these filtered .gtf files\n",
        "2.   Convert to proteins\n",
        "3.   Only retain those from the 450 good genomes corresponding to 44 lineages"
      ],
      "metadata": {
        "id": "Ktsgkp2DncAH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Extract CDS:\n",
        "\n",
        "for i in $(ls ../Ecoli_gtfs/prodigal_gtfs/*prodigal.filtered.gtf | rev | cut -f4- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "cat ../Ecoli_gtfs/\"$i\".filtered.nr.gtf | gtf2bed | bedtools getfasta -s -name -fi /stor/work/Ochman/hassan/MS_Ecoli_ORFans_Ch3/Ecoli_genomes/\"$i\".fasta -bed - > ../Ecoli_cds/\"$i\".filtered.CDS.faa\n",
        "done\n",
        "\n",
        "#Translate to protein:\n",
        "\n",
        "for i in $(ls ../Ecoli_cds/*filtered.CDS.faa | rev | cut -f4- -d \".\" | cut -f1 -d \"/\" | rev)\n",
        "do\n",
        "/stor/work/Ochman/hassan/tools/faTrans -stop ../Ecoli_cds/\"$i\".filtered.CDS.faa ../Ecoli_prot/\"$i\".filtered.prot.faa\n",
        "done\n",
        "\n",
        "#Retain only those proteins annotated in the 450 good genomes, and concatenate them all into one file:\n",
        "\n",
        "for i in $(awk -F '\\t' '($5<45)' /stor/work/Ochman/hassan/Ecoli_pangenome/500_ipp_lineagedesignations.tsv | cut -f1)\n",
        "do\n",
        "sed \"s/>/>\"$i\"@/g\" ../Ecoli_prot/\"$i\".filtered.prot.faa >> Ecoli.all.filtered.prot.faa\n",
        "sed \"s/>/>\"$i\"@/g\" ../Ecoli_cds/\"$i\".filtered.CDS.faa >> Ecoli.all.filtered.CDS.faa\n",
        "done\n",
        "\n",
        "#Number of proteins - 2,066,371"
      ],
      "metadata": {
        "id": "CK1h15zJn3IH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Clustering proteins**\n",
        "\n",
        "The next step(s) involve clustering these 2 million+ proteins into gene families, based on their similarities. We can collapse orthologs and paralogs.\n",
        "\n",
        "Mmseqs2+linclust is a widely adopted and rapid clustering process. We use --cov-mode 0, which requires the alignment to cover 80% of the length of both query and protein. This process is too permissive and will fail to cluster proteins with genuine outgroup hits.\n",
        "\n",
        "To account for this, we pick the longest representative sequence from mmseqs2 results and run an all-vs-all protein blast on them, followed by an additional clustering step with silix."
      ],
      "metadata": {
        "id": "51IrFVgwpSck"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **1. Clustering by MMseqs2**"
      ],
      "metadata": {
        "id": "bPx3c_zkPa30"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#mmseqs + linclust\n",
        "\n",
        "mmseqs createdb Ecoli.all.filtered.prot.faa Ecoli.all.filtered.prot\n",
        "\n",
        "mkdir mmseqs_covmode0\n",
        "cp Ecoli.all.filtered.prot\n",
        "cd mmseqs_covmode0\n",
        "\n",
        "mmseqs search Ecoli.all.filtered.prot Ecoli.all.filtered.prot resultDB tmp --min-seq-id 0.8 -c 0.8 --cov-mode 0\n",
        "mmseqs convertalis Ecoli.all.filtered.prot Ecoli.all.filtered.prot resultDB resultDB.m8\n",
        "mmseqs linclust Ecoli.all.filtered.prot clusterDB tmp --min-seq-id 0.8 -c 0.8 --cov-mode 0\n",
        "mmseqs createtsv Ecoli.all.filtered.prot Ecoli.all.filtered.prot clusterDB Ecoli.all.filtered.prot.clusters.covmode0.tsv\n",
        "\n",
        "#38,797 proteins\n",
        "\n",
        "#At this stage, I'm setting the similarity and query coverage cutoffs pretty high\n",
        "#I don't want to be over-enthusiastic in collapsing actually distinct proteins into the same family\n",
        "#This is just a heuristic step to reduce the number of searches I have to wade through—\n",
        "#...doesn't necessarily reflect \"real\" protein families.\n",
        "#In subsequent phases of this pipeline, if (when) I discover similarities between proteins, we might re-cluster\n",
        "\n",
        "#Extract longest sequence from each cluster\n",
        "#For this purpose, first extract lengths\n",
        "seqkit fx2tab Ecoli.all.filtered.prot.faa | awk -F '\\t' '{print $1,length($2)}' | sort -k1 > Ecoli.all.filtered.prot.lengths.tsv\n",
        "\n",
        "sort -k2 mmseqs_covmode0/Ecoli.all.filtered.prot.clusters.covmode0.tsv | join -1 2 -2 1 - Ecoli.all.filtered.prot.lengths.tsv | #attaching lengths to each gene\n",
        "sort -k2 | awk '{print $0 | \"sort -k2,2 -k3,3n\"}' | awk '!seen[$2]++' | sort -k2 > representative_sequences.interim.tsv #Longest sequence per cluster retained\n",
        "\n",
        "#Make a three-column tsv file which lists longest sequence per cluster in col1, representative sequence picked by mmseqs2 in column3, and genes belonging to that cluster in col2\n",
        "sort -k1 mmseqs_covmode0/Ecoli.all.filtered.prot.clusters.covmode0.tsv | join -1 1 -2 2 -o 1.1 2.2 1.2 2.1 - representative_sequences.interim.tsv | awk '{print $4,$3,$1}' | sed \"s/ /\\t/g\" > Ecoli.all.filtered.prot.clusters.longestrepresentative.tsv\n",
        "\n",
        "#Get CDS and protein sequences\n",
        "seqkit fx2tab Ecoli.all.filtered.prot.faa | sed \"s/\\t$//g\" | sed \"s/^/>/g\" | sed \"s/\\t/\\n/g\" > Ecoli.all.filtered.prot.linear.faa\n",
        "cut -f1 Ecoli.all.filtered.prot.clusters.longestrepresentative.tsv | sort -u | grep --no-group-separator -A1 -F -f - Ecoli.all.filtered.prot.linear.faa > Ecoli.all.filtered.prot.clusters.longestrepresentative.faa\n",
        "cut -f1 Ecoli.all.filtered.prot.clusters.longestrepresentative.tsv | sort -u | grep --no-group-separator -A1 -F -f - Ecoli.all.filtered.CDS.faa > Ecoli.all.filtered.prot.clusters.longestrepresentative.CDS.faa"
      ],
      "metadata": {
        "id": "oM1_f3k9Urqs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Debugging"
      ],
      "metadata": {
        "id": "E5HeSHbkTTyx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "#Footnote: I used a different approach to assign longest sequence per cluster and the results varied in case of 394 proteins\n",
        "#Debugging this might not be worth it, as I doubt it'll have a large footprint on subsequent steps\n",
        "\n",
        "#Assign cluster numbers to each representative sequence:\n",
        "sort -k1 mmseqs_covmode0/Ecoli.all.filtered.prot.clusters.covmode0.tsv | awk 'BEGIN{c=0;prev=\"\"} $1!=prev{c++;prev=$1} {print \"cluster_\" c, $0}' | awk '{print $1\"\\t\"$2\"\\t\"$3}' | sort -k3 > Ecoli.all.filtered.prot.clusters.covmode0.interim.tsv\n",
        "#Get all protein lengths:\n",
        "seqkit fx2tab Ecoli.all.filtered.prot.linear.faa | awk -F '\\t' '{print $1,length($2)}' | sort -k1 | sed \"s/ /\\t/g\" > genelengths_interim.tsv\n",
        "#Join these:\n",
        "join -1 1 -2 3 genelengths_interim.tsv Ecoli.all.filtered.prot.clusters.covmode0.interim.tsv | sort -k4,4 -k2,2 | awk '!seen[$4]++' | cut -f1,4 -d \" \" | sed \"s/ /\\t/g\" | sort -k2 > 1st_longest_2nd_OG.tsv\n",
        "sort -k2 Ecoli.all.filtered.prot.clusters.covmode0.interim.tsv | join -1 2 -2 2 - 1st_longest_2nd_OG.tsv | awk '{print $4\"\\t\"$3\"\\t\"$1}' | sort -k2 | sed \"1s/^/longestrep\\tgene\\tOGcluster\\n/g\" > Ecoli.all.filtered.prot.clusters.longestrepresentative.secondshot.tsv\n",
        "#compare this with previous:\n",
        "cat Ecoli.all.filtered.prot.clusters.longestrepresentative.tsv | sed \"s/\\t/,/g\" | paste - Ecoli.all.filtered.prot.clusters.longestrepresentative.secondshot.tsv | rev | sed \"s/\\t/,/1\" | sed \"s/\\t/,/1\" | rev | awk '($1!=$2)' > weird\n",
        "#isolate cases where the longest rep seq differ:\n",
        "sed \"s/\\t/,/g\" weird  | cut -f1,4 -d ',' | sort -u | cat -n | sed \"s/^ *//g\" | sed \"s/^/group_/g\" | sed \"s/\\t/,/g\" > weird.interim\n",
        "#Hypothesis: second approach picks out longer ones\n",
        "for i in $(cat weird.interim)\n",
        "do\n",
        "echo $i | cut -f2 -d ',' | sed \"s/^/>/g\" | grep -A1 -F -f - Ecoli.all.filtered.prot.linear.faa | seqkit fx2tab | awk '{print length($2)}' > temp1\n",
        "echo $i | cut -f3 -d ',' | sed \"s/^/>/g\" | grep -A1 -F -f - Ecoli.all.filtered.prot.linear.faa | seqkit fx2tab | awk '{print length($2)}' > temp2\n",
        "paste temp1 temp2 | sed \"s/^/\"$i\"\\t/g\" >> weird.lengths.csv\n",
        "rm temp*\n",
        "done"
      ],
      "metadata": {
        "id": "vgka9C7KTISt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **2. Clustering by Silix**"
      ],
      "metadata": {
        "id": "gabGGNoTPiFL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Let's cluster mmseqs2 clusters with silix:\n",
        "\n",
        "#Run an all-vs-all blastp with diamond:\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in Ecoli.all.filtered.prot.clusters.longestrepresentative.faa --db Ecoli.all.filtered.prot.clusters.longestrepresentative\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond blastp -q Ecoli.all.filtered.prot.clusters.longestrepresentative.faa -d Ecoli.all.filtered.prot.clusters.longestrepresentative --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore --ultra-sensitive --out Ecoli.all.filtered.prot.clusters.longestrepresentative.allvall.tsv -k 0 -b8 -c1\n",
        "\n",
        "#Parse the results with silix:\n",
        "#The options -q -2 capture the mmseqs2 logic\n",
        "#-q -2: coverage of the longer sequence -- meaning both of the sequences need to be covered at 80%\n",
        "#-s 3: identity is calculated by counting identical positions divided by aligned sequence length\n",
        "/stor/work/Ochman/hassan/tools/silix-1.3.0/src/silix -f mmseqs2_cluster_ -i 0.8 -r 0.8 -q -2 -s 3 Ecoli.all.filtered.prot.clusters.longestrepresentative.faa Ecoli.all.filtered.prot.clusters.longestrepresentative.allvall.tsv > mmseqs2_recluster_silix_output.tsv\n",
        "\n",
        "#33,245 clusters—14.3% reduction\n",
        "\n",
        "#Now to make a replacements file: a two-column file clustering file that says which sequence is to be replaced with which\n",
        "#For each silix cluster, all genes in the cluster are to be replaced with the longest sequence in that cluster\n",
        "#Extract the sequences in each cluster and place them in files\n",
        "cut -f1 mmseqs2_recluster_silix_output.tsv | sort | uniq -c | grep -v \" 1 \" | rev | cut -f1 -d \" \" | rev | sed \"s/$/\\t/g\" | grep -F -f - mmseqs2_recluster_silix_output.tsv | sed \"s/\\t/,/g\" > mmseqs2_recluster_silix_output.interim.tsv\n",
        "for i in $(cat mmseqs2_recluster_silix_output.interim.tsv | cut -f1 -d \",\")\n",
        "do\n",
        "echo $i | sed \"s/$/,/g\" | grep -F -f - mmseqs2_recluster_silix_output.interim.tsv | cut -f2 -d ',' | sed \"s/^/>/g\" | grep --no-group-separator -A1 -F -f - ../Ecoli.all.filtered.prot.clusters.longestrepresentative.faa > \"$i\".seqs.faa\n",
        "done\n",
        "\n",
        "#Now to set up two-column replacements, where the sequences in column 2 are to be replaced by those of column1:\n",
        "#Column1 contains the longest seq per cluster\n",
        "for i in $(ls mmseqs2_cluster_*.seqs.faa)\n",
        "do\n",
        "temp=$(seqkit fx2tab $i | awk -F '\\t' '{print $1,length($2)}' | sort -nrk2 | head -1 | cut -f1 -d \" \")\n",
        "grep \"^>\" $i | tr -d \">\" | sed \"s/^/\"$temp\"\\t/g\" >> replacements.tsv\n",
        "done\n",
        "\n"
      ],
      "metadata": {
        "id": "FuQriN8rUvdP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Debugging"
      ],
      "metadata": {
        "id": "jWbBEawiZ06A"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Why is there a reduction in gene family number when silix is run on mmseqs2 results?\n",
        "#Are the genes placed in silix clusters actually similar?\n",
        "\n",
        "#Let's look at the sequences that are being clustered together at this stage\n",
        "cut -f1 mmseqs2_recluster_silix_output.tsv | sort | uniq -c | grep -v \" 1 \" | rev | cut -f1 -d \" \" | rev | sed \"s/$/\\t/g\" | grep -F -f - mmseqs2_recluster_silix_output.tsv | sed \"s/\\t/,/g\" > mmseqs2_recluster_silix_output.interim.tsv\n",
        "\n",
        "for i in $(cat mmseqs2_recluster_silix_output.interim.tsv | cut -f1 -d \",\")\n",
        "do\n",
        "echo $i | sed \"s/$/,/g\" | grep -F -f - mmseqs2_recluster_silix_output.interim.tsv | cut -f2 -d ',' | sed \"s/^/>/g\" | grep --no-group-separator -A1 -F -f - ../Ecoli.all.filtered.prot.clusters.longestrepresentative.faa > \"$i\".seqs.faa\n",
        "done\n",
        "\n",
        "#When I check the sequences placed in the same cluster, they are similar\n",
        "#So silix is indeed picking up similar sequences that mmseqs2 missed\n",
        "\n",
        "#Could this have been an artifact of picking the longest sequences?\n",
        "#To check, let's re-run silix using the representative sequences suggested by mmseqs2:\n",
        "makedir OG_repseq_clustering\n",
        "cd OG_repseq_clustering\n",
        "cut -f1 ../mmseqs_covmode0/Ecoli.all.filtered.prot.clusters.covmode0.tsv | sed \"s/^/>/g\" | grep --no-group-separator -A1 -F -f - ../Ecoli.all.filtered.prot.linear.faa > Ecoli.all.filtered.prot.clusters.OGcluster.faa\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond makedb --in Ecoli.all.filtered.prot.clusters.OGcluster.faa --db Ecoli.all.filtered.prot.clusters.OGcluster\n",
        "/stor/work/Ochman/hassan/E.coli_ORFan/E.coli_ORFan_pipeline_8-10/diamond blastp -q Ecoli.all.filtered.prot.clusters.OGcluster.faa -d Ecoli.all.filtered.prot.clusters.OGcluster --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore --ultra-sensitive --out Ecoli.all.filtered.prot.clusters.OGcluster.allvall.tsv -k 0 -b8 -c1\n",
        "/stor/work/Ochman/hassan/tools/silix-1.3.0/src/silix -f mmseqs2_cluster_ -i 0.8 -r 0.8 -q -2 -s 3 Ecoli.all.filtered.prot.clusters.OGcluster.faa Ecoli.all.filtered.prot.clusters.OGcluster.allvall.tsv > mmseqs2_recluster_silix_output.tsv\n",
        "\n",
        "#33818 clusters—12.8%. Not as drastic a reduction as before, but still a reduction. So the Silix-mediated reduction of gene family number is not an artifact\n"
      ],
      "metadata": {
        "id": "aSiIFBv-Z7K3"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}